diff --git a/src/kdeinit/CMakeLists.txt b/src/kdeinit/CMakeLists.txt
index 2a3fecf..0d4a90f 100644
--- a/src/kdeinit/CMakeLists.txt
+++ b/src/kdeinit/CMakeLists.txt
@@ -50,7 +50,7 @@ if (NOT WIN32)
   target_link_libraries(kdeinit5_wrapper  ${KINIT_SOCKET_LIBRARY} Qt5::Core)
 
   if (NOT X11_FOUND)
-    target_compile_definitions(kdeinit5_wrapper -DNO_DISPLAY)
+    target_compile_definitions(kdeinit5_wrapper PUBLIC -DNO_DISPLAY)
   endif ()
 
   install(TARGETS kdeinit5_wrapper ${INSTALL_TARGETS_DEFAULT_ARGS} )
diff --git a/src/kdeinit/kinit.cpp b/src/kdeinit/kinit.cpp
index 8c7e53d..7c4f684 100644
--- a/src/kdeinit/kinit.cpp
+++ b/src/kdeinit/kinit.cpp
@@ -121,6 +121,8 @@ static QByteArray displayEnvVarName()
     return "MAC_DISPLAY";
 #elif defined(Q_OS_WIN)
     return "WIN_DISPLAY";
+#else
+    return 0;
 #endif
 }
 
