diff --git a/src/core/CMakeLists.txt b/src/core/CMakeLists.txt
index 9d0fa27..144fc40 100644
--- a/src/core/CMakeLists.txt
+++ b/src/core/CMakeLists.txt
@@ -1,7 +1,9 @@
 project(KIOCore)
 
+include (CheckLibraryExists)
 include (ConfigureChecks.cmake)
 
+
 configure_file(config-kiocore.h.cmake ${CMAKE_CURRENT_BINARY_DIR}/config-kiocore.h )
 
 configure_file(config-kmountpoint.h.cmake ${CMAKE_CURRENT_BINARY_DIR}/config-kmountpoint.h)
