diff --git src/platformtheme/kdeplatformsystemtrayicon.h src/platformtheme/kdeplatformsystemtrayicon.h
index 6ceaa43..3c1bbf7 100644
--- src/platformtheme/kdeplatformsystemtrayicon.h
+++ src/platformtheme/kdeplatformsystemtrayicon.h
@@ -76,6 +76,7 @@ public:
     void setText(const QString &text) Q_DECL_OVERRIDE;
     void setVisible(bool isVisible) Q_DECL_OVERRIDE;
     quintptr tag() const Q_DECL_OVERRIDE;
+    void setIconSize(int size) /*Q_DECL_OVERRIDE*/; //enable when depending on Qt5.4
 
     QAction *action() const;
 
diff --git src/platformtheme/kdeplatformsystemtrayicon.cpp src/platformtheme/kdeplatformsystemtrayicon.cpp
index 3ada7d2..407e471 100644
--- src/platformtheme/kdeplatformsystemtrayicon.cpp
+++ src/platformtheme/kdeplatformsystemtrayicon.cpp
@@ -24,6 +24,7 @@
 #include <QMenu>
 #include <QRect>
 #include <QApplication>
+#include <QDebug>
 
 #if (QT_VERSION >= QT_VERSION_CHECK(5, 3, 0))
 SystemTrayMenu::SystemTrayMenu()
@@ -212,6 +213,11 @@ void SystemTrayMenuItem::setVisible(bool isVisible)
     m_action->setVisible(isVisible);
 }
 
+void SystemTrayMenuItem::setIconSize(int size)
+{
+    qWarning() << "Cannot set the system tray menu item icon size to " << size;
+}
+
 quintptr SystemTrayMenuItem::tag() const
 {
     return m_tag;
