diff --git a/view.c b/view.c
index 3f3b0ed..ea5db3e 100644
--- a/view.c
+++ b/view.c
@@ -235,10 +235,10 @@ view_map(struct cg_view *view, struct wlr_surface *surface)
 	view->wlr_surface = surface;
 
 	struct wlr_subsurface *subsurface;
-	wl_list_for_each (subsurface, &view->wlr_surface->subsurfaces_below, parent_link) {
+	wl_list_for_each (subsurface, &view->wlr_surface->current.subsurfaces_below, current.link) {
 		subsurface_create(view, subsurface);
 	}
-	wl_list_for_each (subsurface, &view->wlr_surface->subsurfaces_above, parent_link) {
+	wl_list_for_each (subsurface, &view->wlr_surface->current.subsurfaces_above, current.link) {
 		subsurface_create(view, subsurface);
 	}
 
