=== modified file 'CMakeLists.txt'
--- CMakeLists.txt	2014-06-21 14:12:07 +0000
+++ CMakeLists.txt	2014-06-23 00:05:13 +0000
@@ -97,6 +97,7 @@
 
 # check for EGL
 set (with_egl no)  # although EGL can be used with X (replacing GLX), it requires drivers that support DRI2, so most of the time GLX works better; so it's disabled by default for now; later, it will be required for Wayland, and we'll have to decide whether we compile with both, or whether we load one of them as a plug-in at run-time. It may depend on EGL having been built to support a given graphic target.
+enable_if_not_defined (enable-egl-support) 
 if (enable-egl-support)
 	pkg_check_modules ("EGL" "egl")
 	if (EGL_FOUND)

