diff --git a/hooks/casper b/hooks/casper
index f377849..52904ba 100755
--- a/hooks/casper
+++ b/hooks/casper
@@ -92,3 +92,4 @@ if [ -z "\$BOOT" ]; then
 fi
 EOF
 fi
+ln -s /bin/udevadm "${DESTDIR}/sbin/udevadm"
diff --git a/scripts/casper-bottom/15autologin b/scripts/casper-bottom/15autologin
index d6c75e6..dcaa8a8 100755
--- a/scripts/casper-bottom/15autologin
+++ b/scripts/casper-bottom/15autologin
@@ -99,4 +99,9 @@ Session=$sddm_session
 EOF
 fi
 
+if [ -x /root/usr/bin/waylandloginmanager ]; then
+    echo "autologin-user=$USERNAME" >> /root/etc/waylandloginmanager.conf
+    echo "wlmlivemode=1" >> /root/etc/waylandloginmanager.conf
+fi
+
 log_end_msg
diff --git a/scripts/casper-bottom/35limit_coredumpctl_usage b/scripts/casper-bottom/35limit_coredumpctl_usage
new file mode 100755
index 0000000..2b1986a
--- /dev/null
+++ b/scripts/casper-bottom/35limit_coredumpctl_usage
@@ -0,0 +1,19 @@
+#!/bin/sh
+
+PREREQ=""
+
+prereqs()
+{
+       echo "$PREREQ"
+}
+
+case $1 in
+# get pre-requisites
+prereqs)
+       prereqs
+       exit 0
+       ;;
+esac
+
+echo "MaxUse=256M" >> /root/etc/systemd/coredump.conf
+
