diff --git a/src/dolphincontextmenu.cpp b/src/dolphincontextmenu.cpp
--- a/src/dolphincontextmenu.cpp
+++ b/src/dolphincontextmenu.cpp
@@ -359,6 +359,12 @@
 
     addSeparator();
 
+    // Insert 'Sort By' and 'View Mode'
+    addAction(m_mainWindow->actionCollection()->action(QStringLiteral("sort")));
+    addAction(m_mainWindow->actionCollection()->action(QStringLiteral("view_mode")));
+
+    addSeparator();
+
     QAction* pasteAction = createPasteAction();
     addAction(pasteAction);
     addSeparator();

