diff --git a/src/lib/ecore_drm/ecore_drm_device.c b/src/lib/ecore_drm/ecore_drm_device.c
index ca560340dc..6ee71b1134 100644
--- a/src/lib/ecore_drm/ecore_drm_device.c
+++ b/src/lib/ecore_drm/ecore_drm_device.c
@@ -230,7 +230,7 @@ ecore_drm_device_find(const char *name, const char *seat)
 
         if ((seat) && (strcmp(seat, devseat))) 
           goto cont;
-        else if (strcmp(devseat, "seat0"))
+        else if ((!seat) && strcmp(devseat, "seat0"))
           goto cont;
 
         devparent = eeze_udev_syspath_get_parent_filtered(device, "pci", NULL);
diff --git a/src/lib/ecore_drm2/ecore_drm2_device.c b/src/lib/ecore_drm2/ecore_drm2_device.c
index 21963c98fb..27489990a2 100644
--- a/src/lib/ecore_drm2/ecore_drm2_device.c
+++ b/src/lib/ecore_drm2/ecore_drm2_device.c
@@ -116,7 +116,7 @@ _drm2_device_find(Elput_Manager *em, const char *seat)
 
         if ((seat) && (strcmp(seat, dseat)))
           goto cont;
-        else if (strcmp(dseat, "seat0"))
+        else if ((!seat) && strcmp(dseat, "seat0"))
           goto cont;
 
         fd = elput_manager_open(em, dpath, -1);
diff --git a/src/lib/elput/elput_logind.c b/src/lib/elput/elput_logind.c
index 842de225e8..8c76fac5d1 100644
--- a/src/lib/elput/elput_logind.c
+++ b/src/lib/elput/elput_logind.c
@@ -523,15 +523,18 @@ _logind_connect(Elput_Manager **manager, const char *seat, unsigned int tty)
         goto seat_err;
      }
 
-   if (!_logind_session_vt_get(em->sid, &em->vt_num))
+   if (seat == "seat0")
      {
-        ERR("Could not get session vt");
-        goto vt_err;
-     }
-   else if ((tty > 0) && (em->vt_num != tty))
-     {
-        ERR("Requested VT %u differs from session VT %u", tty, em->vt_num);
-        goto vt_err;
+        if (!_logind_session_vt_get(em->sid, &em->vt_num))
+          {
+             ERR("Could not get session vt");
+             goto vt_err;
+          }
+        else if ((tty > 0) && (em->vt_num != tty))
+         {
+             ERR("Requested VT %u differs from session VT %u", tty, em->vt_num);
+             goto vt_err;
+         }
      }
 
    free(s);
