diff --git a/CMakeLists.txt b/CMakeLists.txt
index 1a19ad7..131dad8 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -431,7 +431,7 @@ set(PCSC_FEATURE_TYPE "OPTIONAL")
 set(PCSC_FEATURE_PURPOSE "smart card")
 set(PCSC_FEATURE_DESCRIPTION "smart card device redirection")
 
-set(FFMPEG_FEATURE_TYPE "RECOMMENDED")
+set(FFMPEG_FEATURE_TYPE "DISABLED")
 set(FFMPEG_FEATURE_PURPOSE "multimedia")
 set(FFMPEG_FEATURE_DESCRIPTION "multimedia redirection, audio and video playback")
 
diff --git a/winpr/CMakeLists.txt b/winpr/CMakeLists.txt
index 99a6d41..c95f8cd 100644
--- a/winpr/CMakeLists.txt
+++ b/winpr/CMakeLists.txt
@@ -71,9 +71,9 @@ add_subdirectory(include)
 
 add_subdirectory(libwinpr)
 
-if(NOT ANDROID AND NOT IOS)
+#if(NOT ANDROID AND NOT IOS)
 	add_subdirectory(tools)
-endif()
+#endif()
 
 # Exporting
 
