diff --git a/gedit/meson.build b/gedit/meson.build
index a960e6cee..8d87f60b1 100644
--- a/gedit/meson.build
+++ b/gedit/meson.build
@@ -209,13 +209,13 @@ python3.install_sources(
 )
 
 # Vala API
-libgedit_vapi = gnome.generate_vapi(
-  'gedit',
-  sources: libgedit_gir[0],
-  packages: ['gio-2.0', 'atk', 'gdk-3.0', 'gtk+-3.0', 'gtksourceview-4'],
-  install: true,
-  install_dir: get_option('datadir') / 'vala/vapi',
-)
+#libgedit_vapi = gnome.generate_vapi(
+#  'gedit',
+#  sources: libgedit_gir[0],
+#  packages: ['gio-2.0', 'atk', 'gdk-3.0', 'gtk+-3.0', 'gtksourceview-4'],
+#  install: true,
+#  install_dir: get_option('datadir') / 'vala/vapi',
+#)
 
 libgedit_dep = declare_dependency(
   include_directories: root_include_dir,
