diff --git a/CMakeLists.txt b/CMakeLists.txt
index 7f6bb7fd..7caeb2ae 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -101,14 +101,15 @@ endif()
 
 # Installation paths
 if(UNIX AND SYSTEM_DIR_INSTALL)
+    include(GNUInstallDirs)
     # Install into the system (/usr/bin or /usr/local/bin)
     set(INSTALL_ROOT      "share/${project}")       # /usr/[local]/share/<project>
-    set(INSTALL_CMAKE     "share/${project}/cmake") # /usr/[local]/share/<project>/cmake
+    set(INSTALL_CMAKE     "${CMAKE_INSTALL_LIBDIR}/cmake") # /usr/[local]/lib/<arch>/cmake/<project>
     set(INSTALL_EXAMPLES  "share/${project}")       # /usr/[local]/share/<project>
     set(INSTALL_DATA      "share/${project}")       # /usr/[local]/share/<project>
     set(INSTALL_BIN       "bin")                    # /usr/[local]/bin
-    set(INSTALL_SHARED    "lib")                    # /usr/[local]/lib
-    set(INSTALL_LIB       "lib")                    # /usr/[local]/lib
+    set(INSTALL_SHARED    "${CMAKE_INSTALL_LIBDIR}") # /usr/[local]/lib/<arch>
+    set(INSTALL_LIB       "${CMAKE_INSTALL_LIBDIR}") # /usr/[local]/lib/<arch>
     set(INSTALL_INCLUDE   "include")                # /usr/[local]/include
     set(INSTALL_DOC       "share/doc/${project}")   # /usr/[local]/share/doc/<project>
     set(INSTALL_SHORTCUTS "share/applications")     # /usr/[local]/share/applications
