diff --git a/glycin/src/sandbox.rs b/glycin/src/sandbox.rs
index 2e6d5a1..dd39077 100644
--- a/glycin/src/sandbox.rs
+++ b/glycin/src/sandbox.rs
@@ -297,6 +297,10 @@ impl Sandbox {
             "--ro-bind",
             "/usr",
             "/usr",
+            // Make /opt available as read only
+            "--ro-bind",
+            "/opt",
+            "/opt",
             // Make tmpfs dev available
             "--dev",
             "/dev",
@@ -509,7 +513,7 @@ impl Sandbox {
         //
         // It's not really cleared due to this issue but nothing we can do about this:
         // <https://github.com/flatpak/flatpak/issues/5271>
-        command.env_clear();
+        //command.env_clear();
 
         // Inherit some environment variables
         for key in INHERITED_ENVIRONMENT_VARIABLES {
@@ -541,7 +545,7 @@ impl Sandbox {
     fn no_sandbox_command(&self) -> Command {
         let mut command = Command::new(self.exec());
 
-        command.env_clear();
+        //command.env_clear();
 
         // Inherit some environment variables
         for key in INHERITED_ENVIRONMENT_VARIABLES {
