diff --git a/panels/background/cc-background-item.c b/panels/background/cc-background-item.c
index dc8850b66..50851e17d 100644
--- a/panels/background/cc-background-item.c
+++ b/panels/background/cc-background-item.c
@@ -156,7 +156,7 @@ render_at_size (GnomeBG *bg,
         GdkPixbuf *pixbuf;
 
         pixbuf = gdk_pixbuf_new (GDK_COLORSPACE_RGB, FALSE, 8, width, height);
-        gnome_bg_draw (bg, pixbuf, gdk_screen_get_default (), FALSE);
+        gnome_bg_draw (bg, pixbuf);
 
         return pixbuf;
 }
