diff --git a/subprojects/gvc/gvc-mixer-control.c b/subprojects/gvc/gvc-mixer-control.c
index 21fc233..432a211 100644
--- a/subprojects/gvc/gvc-mixer-control.c
+++ b/subprojects/gvc/gvc-mixer-control.c
@@ -3316,10 +3316,10 @@ gvc_mixer_new_pa_context (GvcMixerControl *self)
                           PA_PROP_APPLICATION_VERSION,
                           PACKAGE_VERSION);
 
-        self->priv->pa_context = pa_context_new_with_proplist (self->priv->pa_api, NULL, proplist);
+//        self->priv->pa_context = pa_context_new_with_proplist (self->priv->pa_api, NULL, proplist);
 
-        pa_proplist_free (proplist);
-        g_assert (self->priv->pa_context);
+//        pa_proplist_free (proplist);
+//        g_assert (self->priv->pa_context);
 }
 
 static void
