diff --git a/src/libgreenisland/compositor.cpp b/src/libgreenisland/compositor.cpp
index 2f6eb39..484f721 100644
--- a/src/libgreenisland/compositor.cpp
+++ b/src/libgreenisland/compositor.cpp
@@ -67,7 +67,7 @@ QString Compositor::s_fixedShell;
 
 Compositor::Compositor(QObject *parent)
     : QObject(parent)
-    , QWaylandQuickCompositor(0, WindowManagerExtension | TouchExtension | HardwareIntegrationExtension | SubSurfaceExtension)
+    , QWaylandQuickCompositor(getenv("WAYLAND_GUEST"), WindowManagerExtension | TouchExtension | HardwareIntegrationExtension | SubSurfaceExtension)
     , d_ptr(new CompositorPrivate(this))
 {
     Q_D(Compositor);
