diff --git a/src/launcher/processcontroller.cpp b/src/launcher/processcontroller.cpp
index dcccf17..aa65a63 100644
--- a/src/launcher/processcontroller.cpp
+++ b/src/launcher/processcontroller.cpp
@@ -138,7 +138,7 @@ void ProcessController::startCompositor()
                                    << QStringLiteral("--socket=") + m_compositorSocket);
 
         env.insert(QStringLiteral("QT_QPA_PLATFORM"), QStringLiteral("wayland"));
-        env.insert(QStringLiteral("WAYLAND_DISPLAY"), m_fullScreenShellSocket);
+        env.insert(QStringLiteral("WAYLAND_GUEST"), m_fullScreenShellSocket);
         if (qEnvironmentVariableIsSet("DISPLAY") && !isFullScreenShellEnabled())
             env.insert(QStringLiteral("QT_XCB_GL_INTEGRATION"), QStringLiteral("xcb_egl"));
         m_compositor->setProcessEnvironment(env);
