diff --git a/plugins/egldeviceintegration/wayland/CMakeLists.txt b/plugins/egldeviceintegration/wayland/CMakeLists.txt
index 79e8b88..1dc4b4f 100644
--- a/plugins/egldeviceintegration/wayland/CMakeLists.txt
+++ b/plugins/egldeviceintegration/wayland/CMakeLists.txt
@@ -1,6 +1,8 @@
+find_package(Qt5PlatformSupport REQUIRED)
 include_directories(
     ${CMAKE_BINARY_DIR}/headers
     ${CMAKE_BINARY_DIR}/headers/GreenIsland
+    ${Qt5PlatformSupport_INCLUDE_DIR}
     ${Qt5Core_PRIVATE_INCLUDE_DIRS}
     ${Qt5Gui_PRIVATE_INCLUDE_DIRS}
 )
