diff --git a/session/compositorlauncher.cpp b/session/compositorlauncher.cpp
index b19f480..66eae0a 100644
--- a/session/compositorlauncher.cpp
+++ b/session/compositorlauncher.cpp
@@ -103,6 +103,10 @@ void CompositorLauncher::start()
         m_compositor->setSocketName(m_socketName);
     }
 
+   if (m_mode == WaylandMode) {
+	m_socketName = getenv("WAYLAND_GUEST");
+        m_compositor->setSocketName(m_socketName);
+   }
     // Summary
     printSummary();
 
@@ -249,6 +253,9 @@ QStringList CompositorLauncher::compositorArgs() const
     case NestedMode:
         args << QStringLiteral("--socket=%1").arg(m_compositor->socketName());
         break;
+    case WaylandMode:
+        args << QStringLiteral("--socket=%1").arg(m_compositor->socketName());
+        break;
     default:
         break;
     }
