diff --git a/kioclient/kioclient.cpp b/kioclient/kioclient.cpp
index af23222..8a88b37 100644
--- a/kioclient/kioclient.cpp
+++ b/kioclient/kioclient.cpp
@@ -34,6 +34,9 @@
 #include <KAboutData>
 #include <KLocalizedString>
 
+#include <QtCore/qurl.h>
+#include "kopenwithdialog.h"
+#include <QWidget>
 #include <QDBusConnection>
 #include <QCommandLineParser>
 #include <QFileDialog>
@@ -63,6 +66,8 @@ static QList<QUrl> makeUrls(const QStringList& urlArgs)
 static void usage()
 {
     puts(i18n("\nSyntax:\n").toLocal8Bit().constData());
+    puts(i18n("  kioclient5 appmenu\n"
+              "            # Opens an application launcher.\n\n").toLocal8Bit().constData());
     puts(i18n("  kioclient5 openProperties 'url'\n"
               "            # Opens a properties menu\n\n").toLocal8Bit().constData());
     puts(i18n("  kioclient5 exec 'url' ['mimetype']\n"
@@ -356,6 +361,15 @@ bool ClientApp::doIt(const QCommandLineParser& parser)
                              argc == 3 ? parser.positionalArguments().last() : QString(),
                              true );
     }
+    else if ( command == QLatin1String("appmenu") )
+    {
+      QList<QUrl> list;
+      KOpenWithDialog dlg(list, i18n("Select an Application to run..."), QString(), 0);
+      dlg.hideRunInTerminal();
+      if (dlg.exec()) {
+          KRun::run(*dlg.service(), list, 0);
+      }
+    }
 #endif
     else if ( command == QLatin1String("download") )
     {
