diff --git a/kioclient/kioclient.cpp b/kioclient/kioclient.cpp
index af23222..5ccf80f 100644
--- a/kioclient/kioclient.cpp
+++ b/kioclient/kioclient.cpp
@@ -22,6 +22,8 @@
 
 #include <KIO/CopyJob>
 #include <KIO/DeleteJob>
+#include <KIO/ApplicationLauncherJob>
+#include <KIO/JobUiDelegate>
 #include <kio/listjob.h>
 #include <kio/transferjob.h>
 #ifndef KIOCORE_ONLY
@@ -34,6 +36,9 @@
 #include <KAboutData>
 #include <KLocalizedString>
 
+#include <QtCore/qurl.h>
+#include "kopenwithdialog.h"
+#include <QWidget>
 #include <QDBusConnection>
 #include <QCommandLineParser>
 #include <QFileDialog>
@@ -356,6 +361,18 @@ bool ClientApp::doIt(const QCommandLineParser& parser)
                              argc == 3 ? parser.positionalArguments().last() : QString(),
                              true );
     }
+    else if ( command == QLatin1String("appmenu") )
+    {
+      QList<QUrl> list;
+      KOpenWithDialog dlg(list, i18n("Select an Application to run..."), QString(), 0);
+      dlg.hideRunInTerminal();
+      if (dlg.exec()) {
+          KIO::ApplicationLauncherJob *job = new KIO::ApplicationLauncherJob(dlg.service());
+          job->setUrls(list);
+          job->start();
+          return m_ok;
+      }
+    }
 #endif
     else if ( command == QLatin1String("download") )
     {
