diff --git a/src/kselectionproxymodel.cpp b/src/kselectionproxymodel.cpp
index 6774b75..a241728 100644
--- a/src/kselectionproxymodel.cpp
+++ b/src/kselectionproxymodel.cpp
@@ -1344,7 +1344,7 @@ QModelIndex KSelectionProxyModelPrivate::mapParentFromSource(const QModelIndex &
     return m_mappedParents.leftToRight(sourceParent);
 }
 
-#ifndef QT_NO_DEBUG
+//#ifndef QT_NO_DEBUG
 static bool indexIsValid(bool startWithChildTrees, int row, const QList<QPersistentModelIndex> &rootIndexList, const SourceIndexProxyRowMapping &mappedFirstChildren)
 {
     if (!startWithChildTrees) {
@@ -1365,7 +1365,7 @@ static bool indexIsValid(bool startWithChildTrees, int row, const QList<QPersist
     }
     return true;
 }
-#endif
+//#endif
 
 QModelIndex KSelectionProxyModelPrivate::createTopLevelIndex(int row, int column) const
 {
