diff --git a/composite.cpp b/composite.cpp
index ef8488f..ad26199 100644
--- a/composite.cpp
+++ b/composite.cpp
@@ -240,9 +240,16 @@ void Compositor::slotCompositingOptionsInitialized()
     }
 #ifdef KWIN_HAVE_XRENDER_COMPOSITING
     case XRenderCompositing:
-        qCDebug(KWIN_CORE) << "Initializing XRender compositing";
-        m_scene = SceneXrender::createScene(this);
-        break;
+        if (! kwinApp()->shouldUseWaylandForCompositing())
+        {
+            qCDebug(KWIN_CORE) << "Initializing XRender compositing";
+            m_scene = SceneXrender::createScene(this);
+            break;
+        } else {
+           qCDebug(KWIN_CORE) << "Initializing QPainter compositing";
+           m_scene = SceneQPainter::createScene(this);
+           break;
+        }
 #endif
     case QPainterCompositing:
         qCDebug(KWIN_CORE) << "Initializing QPainter compositing";
@@ -272,9 +279,9 @@ void Compositor::slotCompositingOptionsInitialized()
             cm_selection->release();
         }
         if (kwinApp()->platform()->requiresCompositing()) {
-            qCCritical(KWIN_CORE) << "The used windowing system requires compositing";
-            qCCritical(KWIN_CORE) << "We are going to quit KWin now as it is broken";
-            qApp->quit();
+            //qCCritical(KWIN_CORE) << "The used windowing system requires compositing";
+            //qCCritical(KWIN_CORE) << "We are going to quit KWin now as it is broken";
+            //qApp->quit();
         }
         return;
     }

     
     

