--- a/build/moz.configure/rust.configure
+++ b/build/moz.configure/rust.configure
@@ -415,6 +415,12 @@
         if len(narrowed) == 1:
             return narrowed[0].rust_target

+        # The autoconf "pc" vendor corresponds to Rust's "unknown" vendor.
+        if host_or_target.vendor == "pc":
+            narrowed = [c for c in candidates if c.target.vendor == "unknown"]
+            if len(narrowed) == 1:
+                return narrowed[0].rust_target
+
         return None

     rustc_target = find_candidate(candidates)

diff --git a/js/src/js-config.h.in b/js/src/js-config.h.in
index 8372623..498ea96 100644
--- a/js/src/js-config.h.in
+++ b/js/src/js-config.h.in
@@ -68,4 +68,8 @@
 #undef MOZJS_MAJOR_VERSION
 #undef MOZJS_MINOR_VERSION
 
+/* Controls API in UniquePtrExtensions.h. */
+#undef XP_UNIX
+#undef XP_WIN
+
 #endif /* js_config_h */
