diff --git a/arch/x86/kernel/setup.c b/arch/x86/kernel/setup.c
index 05c5aa951da7..be68dd3f49eb 100644
--- a/arch/x86/kernel/setup.c
+++ b/arch/x86/kernel/setup.c
@@ -1159,6 +1159,11 @@ void __init setup_arch(char **cmdline_p)
 
 	e820__setup_pci_gap();
 
+#ifdef CONFIG_NULL_TTY_CONSOLE
+	if (!strstr(boot_command_line, "console="))
+		add_preferred_console("ttynull", 0, NULL);
+#endif
+
 #ifdef CONFIG_VT
 #if defined(CONFIG_VGA_CONSOLE)
 	if (!efi_enabled(EFI_BOOT) || (efi_mem_type(0xa0000) != EFI_CONVENTIONAL_MEMORY))
diff --git a/drivers/tty/Kconfig b/drivers/tty/Kconfig
index a45d423ad10f..a8080732589a 100644
--- a/drivers/tty/Kconfig
+++ b/drivers/tty/Kconfig
@@ -387,6 +387,17 @@ config NULL_TTY
 
 	  If unsure, say N.
 
+config NULL_TTY_CONSOLE
+        bool "Supports /dev/ttynull as a console automatically"
+        depends on NULL_TTY && !VT_CONSOLE
+	help
+	  Say Y here if you want a NULL TTY to be /dev/console automatically.
+
+	  This is useful for replacing where VT_CONSOLE was previously enabled.
+	  /dev/ttynull backs /dev/console if no console= options are specified
+
+	  If unsure, say N.
+
 config VCC
 	tristate "Sun Virtual Console Concentrator"
 	depends on SUN_LDOMS
