diff --git a/src/plugins/decorations/material/materialdecoration.cpp b/src/plugins/decorations/material/materialdecoration.cpp
index 3c236e37..17935f43 100644
--- a/src/plugins/decorations/material/materialdecoration.cpp
+++ b/src/plugins/decorations/material/materialdecoration.cpp
@@ -167,17 +167,16 @@ void QWaylandMaterialDecoration::paint(QPaintDevice *device)
     // Title bar
     int radius = waylandWindow()->isMaximized() ? 0 : dp(3);
     QPainterPath roundedRect;
-    roundedRect.addRoundedRect(margins().left(), margins().top() - TITLE_BAR_HEIGHT,
-                               frameGeometry.width() - margins().left() - margins().right(), TITLE_BAR_HEIGHT + radius * 2,
+    roundedRect.addRoundedRect(0, 0, frameGeometry.width(), margins().top() * 1.5,
                                radius, radius);
     p.fillPath(roundedRect.simplified(), m_backgroundColor);
 
-    // Borders (transparent so the border is not noticeable)
+    // Borders
     QPainterPath borderPath;
     borderPath.addRect(0, margins().top(), margins().left(), frameGeometry.height() - margins().top());
     borderPath.addRect(0, frameGeometry.height() - margins().bottom(), frameGeometry.width(), margins().bottom());
     borderPath.addRect(frameGeometry.width() - margins().right(), margins().top(), margins().right(), frameGeometry.height() - margins().bottom());
-    p.fillPath(borderPath, Qt::transparent);
+    p.fillPath(borderPath, m_backgroundColor);
 
     // Window title
     QString windowTitleText = window()->title();
