diff --git a/src/egl/drivers/dri2/platform_drm.c b/src/egl/drivers/dri2/platform_drm.c
index d0ab172d659..a72f1eb1e76 100644
--- a/src/egl/drivers/dri2/platform_drm.c
+++ b/src/egl/drivers/dri2/platform_drm.c
@@ -695,9 +695,9 @@ dri2_initialize_drm(_EGLDriver *drv, _EGLDisplay *disp)
    struct gbm_device *gbm;
    const char *err;
 
-   /* Not supported yet */
-   if (disp->Options.ForceSoftware)
-      return EGL_FALSE;
+   ///* Not supported yet */
+   //if (disp->Options.ForceSoftware)
+   //   return EGL_FALSE;
 
    dri2_dpy = calloc(1, sizeof *dri2_dpy);
    if (!dri2_dpy)
@@ -735,7 +735,7 @@ dri2_initialize_drm(_EGLDriver *drv, _EGLDisplay *disp)
    dev = _eglAddDevice(dri2_dpy->fd, false);
    if (!dev) {
       err = "DRI2: failed to find EGLDevice";
-      goto cleanup;
+      //goto cleanup;
    }
 
    disp->Device = dev;
