diff --git a/CMakeLists.txt b/CMakeLists.txt
index 4a6e0d8..272b16b 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -1,7 +1,7 @@
 cmake_minimum_required(VERSION 2.8)
 project(orbital)
 
-set(CMAKE_CXX_FLAGS "-Wall -Wextra -Wno-unused-parameter -g -std=c++0x -Werror=return-type")
+set(CMAKE_CXX_FLAGS "-Wall -Wextra -Wno-unused-parameter -g -std=c++0x")
 set(CMAKE_MODULE_PATH "${CMAKE_SOURCE_DIR}/cmake")
 
 include(Wayland)
diff --git a/src/client/CMakeLists.txt b/src/client/CMakeLists.txt
index 28caf60..fbfb0f1 100644
--- a/src/client/CMakeLists.txt
+++ b/src/client/CMakeLists.txt
@@ -11,7 +11,7 @@ function(INSTALL_ELEMENT _sources _dir)
             get_filename_component(_abs_ts ${ts} ABSOLUTE)
             get_filename_component(_name ${ts} NAME_WE)
             set(qm "${CMAKE_CURRENT_BINARY_DIR}/${_dir}/${_name}.qm")
-            add_custom_command(OUTPUT ${qm} COMMAND ${LRELEASE_EXECUTABLE} ARGS -qt5 ${_abs_ts} -qm ${qm} DEPENDS ${_abs} VERBATIM)
+            add_custom_command(OUTPUT ${qm} COMMAND ${LRELEASE_EXECUTABLE} ARGS ${_abs_ts} -qm ${qm} DEPENDS ${_abs} VERBATIM)
 
             list(APPEND ${_sources} "${qm}")
             install(FILES ${qm} DESTINATION share/orbital/${_dir})
