diff --git a/plasma-windowed/CMakeLists.txt b/plasma-windowed/CMakeLists.txt
index ff5a020f7..7de1db3cb 100644
--- a/plasma-windowed/CMakeLists.txt
+++ b/plasma-windowed/CMakeLists.txt
@@ -26,4 +26,5 @@ target_link_libraries(
 
 install(TARGETS plasmawindowed ${KDE_INSTALL_TARGETS_DEFAULT_ARGS})
 #even if hidden, the desktop file is needed anyways for kdbusservice::unique
-install(FILES plasma-windowed.desktop DESTINATION ${KDE_INSTALL_APPDIR})
+configure_file(org.kde.plasmawindowed.desktop.cmake ${CMAKE_CURRENT_BINARY_DIR}/org.kde.plasmawindowed.desktop @ONLY)
+install(FILES ${CMAKE_CURRENT_BINARY_DIR}/org.kde.plasmawindowed.desktop DESTINATION ${KDE_INSTALL_APPDIR})
diff --git a/plasma-windowed/org.kde.plasmawindowed.desktop.cmake b/plasma-windowed/org.kde.plasmawindowed.desktop.cmake
index 09c7994f8..4daf2dcc9 100755
--- a/plasma-windowed/org.kde.plasmawindowed.desktop.cmake
+++ b/plasma-windowed/org.kde.plasmawindowed.desktop.cmake
@@ -101,10 +101,11 @@ Comment[x-test]=xxRun Plasmoids as simple applicationsxx
 Comment[zh_CN]=将小部件当作简单应用程序运行
 Comment[zh_TW]=以簡單應用程式執行 Plasma 小元件
 
-Exec=plasmawindowed %u
+Exec=@CMAKE_INSTALL_PREFIX@/bin/plasmawindowed
 Terminal=false
 Hidden=true
 Icon=plasma
 Type=Application
 X-DBUS-StartupType=Unique
 InitialPreference=1
+X-KDE-Wayland-Interfaces=org_kde_plasma_window_management,org_kde_kwin_keystate,zkde_screencast_unstable_v1,org_kde_plasma_activation_feedback
diff --git a/plasma-windowed/org.kde.plasmawindowed.desktop.cmake b/plasma-windowed/org.kde.plasmawindowed.desktop.cmake
index 4daf2dcc9..5e861b4fc 100755
--- a/plasma-windowed/org.kde.plasmawindowed.desktop.cmake
+++ b/plasma-windowed/org.kde.plasmawindowed.desktop.cmake
@@ -103,7 +103,7 @@ Comment[zh_TW]=以簡單應用程式執行 Plasma 小元件
 
 Exec=@CMAKE_INSTALL_PREFIX@/bin/plasmawindowed
 Terminal=false
-Hidden=true
+NoDisplay=true
 Icon=plasma
 Type=Application
 X-DBUS-StartupType=Unique
