diff --git a/CMakeLists.txt b/CMakeLists.txt
index f82b7f0b76..f8dda17961 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -149,6 +149,12 @@ set_package_properties(ZLIB PROPERTIES DESCRIPTION "Support for gzip compressed
                        TYPE REQUIRED
                       )
 
+find_package(Freetype)
+set_package_properties(Freetype PROPERTIES DESCRIPTION "Software library to render fonts"
+                       URL "https://freetype.org"
+                       TYPE OPTIONAL
+                      )
+
 find_package(Fontconfig)
 set_package_properties(Fontconfig PROPERTIES DESCRIPTION "Font access configuration library"
                        URL "https://www.freedesktop.org/wiki/Software/fontconfig"
diff --git a/kcms/kfontinst/lib/CMakeLists.txt b/kcms/kfontinst/lib/CMakeLists.txt
index 54de2fa83c..1f5cf13a93 100644
--- a/kcms/kfontinst/lib/CMakeLists.txt
+++ b/kcms/kfontinst/lib/CMakeLists.txt
@@ -9,6 +9,7 @@ target_link_libraries(kfontinst
     KF6::I18n
     KF6::KIOCore
     Fontconfig::Fontconfig
+    Freetype::Freetype
     PRIVATE
     Qt::Xml
     Qt::Gui
diff --git a/kcms/kfontinst/lib/Fc.h b/kcms/kfontinst/lib/Fc.h
index 4dd854d4db..88e7fedeef 100644
--- a/kcms/kfontinst/lib/Fc.h
+++ b/kcms/kfontinst/lib/Fc.h
@@ -8,8 +8,10 @@
 #include "Misc.h"
 #include "kfontinst_export.h"
 #include <QUrl>
+#include <fontconfig/fcfreetype.h>
 #include <fontconfig/fontconfig.h>
 
+
 #include "kfontinst_export.h"
 
 #if (FC_VERSION < 20200)
