diff --git a/src/libply-splash-core/ply-device-manager.c b/src/libply-splash-core/ply-device-manager.c
index 40d3334d..40b27f8a 100644
--- a/src/libply-splash-core/ply-device-manager.c
+++ b/src/libply-splash-core/ply-device-manager.c
@@ -385,7 +385,9 @@ create_devices_for_udev_device (ply_device_manager_t *manager,
                         ply_terminal_t *terminal = NULL;
 
                         if (!manager->local_console_managed)
-                                terminal = manager->local_console_terminal;
+                                if (manager->local_console_terminal != NULL &&
+                                    ply_terminal_is_vt (manager->local_console_terminal))
+                                        terminal = manager->local_console_terminal;
 
                         created = create_devices_for_terminal_and_renderer_type (manager,
                                                                                  device_path,
diff --git a/src/main.c b/src/main.c
index 3a987873..d4e4ad84 100644
--- a/src/main.c
+++ b/src/main.c
@@ -156,6 +156,7 @@ static void on_quit (state_t       *state,
                      bool           retain_splash,
                      ply_trigger_t *quit_trigger);
 static bool sh_is_init (state_t *state);
+static bool plymouth_arguments_want_splash (state_t *state);
 static void cancel_pending_delayed_show (state_t *state);
 static void prepare_logging (state_t *state);
 static void dump_debug_buffer_to_file (void);
@@ -872,14 +873,21 @@ plymouth_should_show_default_splash (state_t *state)
 {
         ply_trace ("checking if plymouth should show default splash");
 
+        if (state->should_force_details)
+                return false;
+
+        if (plymouth_arguments_want_splash (state))
+                return true;
+}
+
+static bool
+plymouth_arguments_want_splash (state_t *state)
+{
         const char * const strings[] = {
                 "single", "1", "s", "S", "-S", NULL
         };
         int i;
 
-        if (state->should_force_details)
-                return false;
-
         for (i = 0; strings[i] != NULL; i++) {
                 if (ply_kernel_command_line_has_argument (strings[i])) {
                         ply_trace ("no default splash because kernel command line has option \"%s\"", strings[i]);
@@ -1518,6 +1526,14 @@ static void
 on_escape_pressed (state_t *state)
 {
         ply_trace ("escape key pressed");
+
+        if (state->local_console_terminal != NULL)
+                if (!ply_terminal_is_vt (state->local_console_terminal))
+                        return;
+        else
+                return;
+
+
         if (validate_input (state, "", "\e"))
                 toggle_between_splash_and_details (state);
 }
@@ -1990,6 +2006,8 @@ initialize_environment (state_t *state)
         if (!state->default_tty)
                 if (getenv ("DISPLAY") != NULL && access (PLYMOUTH_PLUGIN_PATH "renderers/x11.so", F_OK) == 0)
                         state->default_tty = "/dev/tty";
+                if (!state->default_tty && plymouth_arguments_want_splash (state))
+                        state->default_tty = find_fallback_tty (state);
         if (!state->default_tty) {
                 if (state->mode == PLY_BOOT_SPLASH_MODE_SHUTDOWN ||
                     state->mode == PLY_BOOT_SPLASH_MODE_REBOOT)
@@ -2350,6 +2368,9 @@ main (int    argc,
         find_system_default_splash (&state);
         find_distribution_default_splash (&state);
 
+        if (plymouth_arguments_want_splash (&state))
+                device_manager_flags |= PLY_DEVICE_MANAGER_FLAGS_IGNORE_SERIAL_CONSOLES;
+
         if (ply_kernel_command_line_has_argument ("plymouth.ignore-serial-consoles") ||
             ignore_serial_consoles == true)
                 device_manager_flags |= PLY_DEVICE_MANAGER_FLAGS_IGNORE_SERIAL_CONSOLES;
