diff --git a/src/core/chromium_gpu_helper.cpp b/src/core/chromium_gpu_helper.cpp
index 73a4da7..547fbc4 100644 (file)
--- a/src/core/chromium_gpu_helper.cpp
+++ b/src/core/chromium_gpu_helper.cpp
@@ -48,7 +48,7 @@
 // GL typedefs cascading through content header includes.
 #include "gpu/command_buffer/service/sync_point_manager.h"
 #include "gpu/command_buffer/service/mailbox_manager.h"
-#include "gpu/command_buffer/service/texture_manager.h"
+#include "gpu/command_buffer/service/texture_base.h"
 
 #include "content/gpu/gpu_child_thread.h"
 #include "gpu/ipc/service/gpu_channel_manager.h"
