diff --git a/Source/WebCore/platform/qt/UserAgentQt.cpp b/Source/WebCore/platform/qt/UserAgentQt.cpp
index 6f1da23..d4850a8 100644
--- a/Source/WebCore/platform/qt/UserAgentQt.cpp
+++ b/Source/WebCore/platform/qt/UserAgentQt.cpp
@@ -67,7 +67,7 @@ String UserAgentQt::standardUserAgent(const String &applicationNameForUserAgent,
 #elif OS(WINDOWS)
             ""
 #else
-            (QGuiApplication::platformName() == QLatin1String("xcb")) ? "X11; " : "Unknown; "
+	    (QGuiApplication::platformName() == QLatin1String("xcb")) ? "X11; " : (QGuiApplication::platformName().contains(QStringLiteral("wayland"))) ? "Wayland; " : "Unknown; "
 #endif
         ));
 


