--- usr/bin/waylandloginmanager	2020-01-03 19:33:49.959040076 -0500
+++ usr/bin/waylandloginmanagerkdialog	2020-01-03 19:35:32.333887458 -0500
@@ -18,9 +18,9 @@
 
 #This script is the WaylandLoginManager it handles a graphical login for the user, and allows the user to start multiple types of Wayland sessions, specified in wsession files, and it supports autologin, mounting of ecryptfs home directories, and user switching. It supports watching the active session until it fails, so that it swiches back to the needed TTY for the user.
 
-#The main server it displays on is refered to as the loginmanagerdisplay. it runs as non root, as well as the zenity dialogs
+#The main server it displays on is refered to as the loginmanagerdisplay. it runs as non root, as well as the kdialog dialogs
 
-#It depends on zenity kbd, weston, and expect
+#It depends on kdialog kbd, weston, and expect
 
 
 #It depends on the line
@@ -114,7 +114,7 @@
      #to override setting any default sessions for a user, when the global one is set.
 
 #/usr/libexec/wayland_login_helpers/*
-  #These scripts are small utilities that send the approriate commands to the waylandloginmanager's FIFO. They need to be configured in the launchers of /etc/loginmanagerdisplay/weston.ini. It's how the user interacts with the waylandloginmanager on the loginmanagerdisplay (other then the zenity dialogs)
+  #These scripts are small utilities that send the approriate commands to the waylandloginmanager's FIFO. They need to be configured in the launchers of /etc/loginmanagerdisplay/weston.ini. It's how the user interacts with the waylandloginmanager on the loginmanagerdisplay (other then the kdialog dialogs)
   #These are optional as the waylandloginmanager displays a selection list with the same commands as well
 #/etc/loginmanagerdisplay/weston.ini
   #Configuration for the loginmanagerdisplay. This should add the launchers in /usr/lib/wayland_login_helpers to the launcher bar, which send commands to the FIFO for weston, or at least prevent the default terminal icon from appearing in the toolbar for security reasons.
@@ -699,7 +699,7 @@
   #If only 1 session should be allowed on seats where logind reports CanMultiSession=no (default is to allow)
   GetConfigKey wlmobeystrictseats "" 0 1 ONLY_CAN_MULTISESSION_SESSION_SWITCH
 
-  #User that the zenity dialogs and the loginmanagerdisplay (weston instance for the waylandloginmanager runs as
+  #User that the kdialog dialogs and the loginmanagerdisplay (weston instance for the waylandloginmanager runs as
   export LOGINMANAGERDISPLAYUSER=waylandloginmanager
   #User that the su test in AuthenticateUser user runs as
   export PROBETESTUSER=waylandloginmanager
@@ -871,7 +871,8 @@
   export DEB_HOST_MULTIARCH=$(dpkg-architecture -qDEB_HOST_MULTIARCH 2>/dev/null)
   export LD_LIBRARY_PATH=/opt/lib/$DEB_HOST_MULTIARCH:/opt/lib:/usr/local/lib/$DEB_HOST_MULTIARCH:/usr/lib/$DEB_HOST_MULTIARCH:/usr/local/lib:/usr/lib
   export XDG_DATA_DIRS=/opt/share:/usr/share
-  export GDK_BACKEND=wayland
+  export QT_QPA_PLATFORM=wayland
+  export QT_STYLE_OVERRIDE=oxygen
   #Set the default path for weston
   export DEFAULTWAYLANDSERVER=$(which weston)
   #Specify where wsession files are found
@@ -889,14 +890,14 @@
   #Determine if this is running on a live instance, if booted with casper
   GetConfigKey boot=casper "" 0 2 ISLIVE
 
-  ZENITYCMD="runuser -u $LOGINMANAGERDISPLAYUSER -m -- zenity"
+  KDIALOGCMD="runuser -u $LOGINMANAGERDISPLAYUSER -m -- kdialog"
   TERMINALCMD="runuser -u $LOGINMANAGERDISPLAYUSER -m -- vte"
   WESTONINFOCMD="runuser -u $LOGINMANAGERDISPLAYUSER -m -- weston-info"
   DIALOGCMD="systemd-run --unit="wlm-ttydialog" --wait -p CollectMode=inactive-or-failed -p TTYPath=/dev/tty63 -p User="$LOGINMANAGERDISPLAYUSER" -p StandardOutput=tty -p StandardError=tty -p StandardInput=tty dialog"
   TTYWIZARDCMD="systemd-run --unit="wlm-ttydialog" --wait -p CollectMode=inactive-or-failed -p TTYPath=/dev/tty63 -p StandardOutput=tty -p StandardError=tty -p StandardInput=tty"
 
-  command -v zenity > /dev/null
-  ZenityCommandFoundResult=$?
+  command -v kdialog > /dev/null
+  KdialogCommandFoundResult=$?
 
   #Detect hardware abilities, first try to see if the system has kernel mode setting, then try framebuffer. if both aren't supported, bring up a text mode dialog
   GetConfigKey wlmforcefbdev "" 0 1 wlmforcefbdev
@@ -1233,23 +1234,14 @@
   done
 }
 
-#This function takes three arguments. The first is the number of seconds to wait, and the second is the string to show in the dialog. it shows a Zenity dialog with a progress bar to count down the specified time, with the string. The third is the seat to display it on
+#This function takes three arguments. The first is the number of seconds to wait, and the second is the string to show in the dialog. it shows a dialog with a progress bar to count down the specified time, with the string. The third is the seat to display it on
 function DialogWait
 {
   GetStartedSeatIndex $3
   SeatFileName=${SeatFileNames[$ReturnSeatIndex]}
   export WAYLAND_DISPLAY=loginmanagerdisplay_"$SeatFileName"
   SetFallbackEnvironmentVariables $3
-  waitseconds=$1
-  displaystring="$2"
-  countseconds=0
-  while [[ $countseconds -lt $waitseconds ]]
-  do
-    percent=$(( $countseconds * 100 / $waitseconds ))
-    echo $percent
-    ((countseconds++))
-    $SLEEPCMD 1
-  done | $ZENITYCMD --width=450 --title="Login Manager" --no-cancel --progress --auto-close --text="$displaystring" 2>/dev/null
+  $KDIALOGCMD --passivepopup "$2" "$1" 2>/dev/null
 }
 
 #This function takes 1 argument, the seat that the loginmanagerdisplay belongs to, and waits for the loginmanagerdisplay to startup
@@ -1868,7 +1860,7 @@
       done
       if [[ $WSESSIONNUMBER == -1 ]]
       then
-        $ZENITYCMD --width=500 --error --title "Session Selection" --text="$DEFAULTWSESSION set as default session for $SessionUser, but was not found, configured correctly, or supported by the hardware in this seat." 2>/dev/null
+        $KDIALOGCMD --title="Session Selection" --error "$DEFAULTWSESSION set as default session for $SessionUser, but was not found, configured correctly, or supported by the hardware in this seat." 2>/dev/null
         WriteToLog "$DEFAULTWSESSION set as default session for $SessionUser, but was not found in $WSESSIONSPATH, configured correctly, or supported by the hardware in this seat, $CurrentHandleSeat ."
         return 1
       fi
@@ -1883,13 +1875,16 @@
         then
           WSESSIONPICKLIST+=$'\n'
         fi
-        WSESSIONPICKLIST+=$element$"#"${WSESSIONARRAY[$((8+element*10))]}"#"${WSESSIONARRAY[$((9+element*10))]}
+        WSESSIONPICKLIST+=$element"#"${WSESSIONARRAY[$((8+element*10))]}$'\t'$'\t'$'\t'${WSESSIONARRAY[$((9+element*10))]}
       done
       WSESSIONPICKLIST=$(echo "$WSESSIONPICKLIST" | sort -t "#" -k 2,2 | sed 's/#/\n/g')
+      IFS=$'\n'
+      WSESSIONPICKLIST=($WSESSIONPICKLIST)
+      unset IFS
       #select the session. Each line is handled as a column. currently there are 9 collumns. So line 1 is is column 1, line 2 is in collumn 2, and line 10 is in collumn 1 again
-      #The zenity dialog then outputs the selected session's lines into the WSESSIONDATA variable. each collumn is a different aspect of the wsessions
-      WSESSIONNUMBER=$(echo "$WSESSIONPICKLIST" | $ZENITYCMD --title="Pick a Session..." --height=450 --width=630 --list --text "Select a Wayland Desktop Environment to use" --column sessionnumber --column Name --column Comment --hide-column=1 --print-column=1 2>/dev/null )
-      #Abort if the user selected cancel on the zenity dialog
+      #The dialog then outputs the selected session's lines into the WSESSIONDATA variable. each collumn is a different aspect of the wsessions
+      WSESSIONNUMBER=$($KDIALOGCMD --geometry=700x300+0+0 --title="Pick a Session..." --menu "Select a Wayland Desktop Environment to use" -- "${WSESSIONPICKLIST[@]}" 2>/dev/null )
+      #Abort if the user selected cancel on the dialog
       CancelOrOK=$?
       if [[ $CancelOrOK != 0 ]]
       then
@@ -1905,7 +1900,7 @@
   else
     unset WSESSIONDATA
     unset WSESSIONNUMBER
-    $ZENITYCMD --width=500 --error --title "Session Selection" --text="No sessions are installed, configured correctly, or supported by the hardware in this seat." 2>/dev/null
+    $KDIALOGCMD --title="Session Selection" --error "No sessions are installed, configured correctly, or supported by the hardware in this seat." 2>/dev/null
     WriteToLog "No sessions are installed in $WSESSIONSPATH, configured correctly, or supported by the hardware in this seat, $CurrentHandleSeat ."
     return 1
   fi
@@ -1963,7 +1958,7 @@
 
 This is the optimal way to run Weston or other Wayland servers."
   fi
-  $ZENITYCMD --title="Information" --no-wrap --info --text="This is a third party fan made distribution!
+  $KDIALOGCMD --title="System Information" --msgbox "This is a third party fan made distribution!
 
 $BACKENDSTRING
 
@@ -1989,12 +1984,6 @@
   ((NumberOfSessions--))
 
   #go thorugh each session, reading the arrays containing information about the sessions. to build the SessionList variable
-  #The session list variable is multiline, and used by a zenity dialog.
-  #line 1 is the session number
-  #line 2 is the TTY for the session
-  #line 3 is the PID of the session
-  #line 4 is the user name
-  #line 5 is the logind session id.
   RunningSeatSessions=($(loginctl show-seat "$1" -p Sessions --value 2>/dev/null))
   while [[ $NumberOfSessions -ge 0 ]]
   do
@@ -2017,11 +2006,7 @@
         then
           SessionList+=$'\n'
         fi
-        SessionList+="$NumberOfSessions"$'\n'
-        SessionList+="${AllSessionIDs[$NumberOfSessions]}"$'\n'
-        SessionList+="${AllSessionPIDs[$NumberOfSessions]}"$'\n'
-        SessionList+="${AllSessionUsers[$NumberOfSessions]}"$'\n'
-        SessionList+="tty${AllSessionTTYs[$NumberOfSessions]}"
+        SessionList+="${AllSessionIDs[$NumberOfSessions]}"$'\n'"${AllSessionIDs[$NumberOfSessions]}"$'\t'"${AllSessionUsers[$NumberOfSessions]}"
       fi
     fi
     ((NumberOfSessions--))
@@ -2032,19 +2017,17 @@
     SessionList+=$'\n'
   fi
   SessionList+="-1"$'\n'
-  SessionList+=$'\n'
-  SessionList+=$'\n'
-  SessionList+="New Session..."$'\n'
-  SessionList+=$'\n'
-
-
+  SessionList+="New Session..."
 
+  IFS=$'\n'
+  SessionList=($SessionList)
+  unset IFS
 
   unset ChangeSession
   #prompt for the list of running sessions, and return the session ID. (which is used in all of the arrays for Session PID, user name, etc)
-  ChangeSession=$(echo "$SessionList"  | sed '$d'| $ZENITYCMD --title="Switch User" --height=450 --list --text "Select a running session to change into" --hide-header --column sessionid --column tty --column pid --column username --column ID --print-column=1 --hide-column=1,3,5 --separator="\n" 2>/dev/null )
+  ChangeSession=$($KDIALOGCMD --title="Switch User" --menu "Select a running session to change into" -- "${SessionList[@]}" 2>/dev/null )
   CancelOrOK=$?
-  #Abort if the user selected cancel on the zenity dialog
+  #Abort if the user selected cancel on the dialog
   if [[ $CancelOrOK != 0 ]]
   then
     exit 0
@@ -2073,7 +2056,7 @@
         #hand over the logind session ID to allow desktop permissions to work.
         loginctl activate $ChangeSessionID &>/dev/null
       else
-        $ZENITYCMD --title="Switch User" --warning --text="The selected session was not found. The session may have quit." 2>/dev/null
+        $KDIALOGCMD --title="Switch User" --msgbox "The selected session was not found. The session may have quit." 2>/dev/null
       fi
     fi
   fi
@@ -2331,26 +2314,30 @@
     then
       UserDisplayName=$UserName
     fi
-    #The display in the zenity dialog should be "Firstname Lastname    (loginname)"
+    #The display in the dialog should be "Firstname Lastname    (loginname)"
     UserDisplayName="$UserDisplayName   (${UserData[1]})" 
     unset IFS
-    #add the data to the USERLISTSTRING that gets sent into zenity. Line 1 is the display name, line 2 is the user login name, and then line 3 is the display name of the next user...
-    USERLISTSTRING+="$UserDisplayName"
-    USERLISTSTRING+=$'\n'
+    #add the data to the USERLISTSTRING that gets sent into kdialog. line 1 is the user login name, Line 2 is the display name, and then line 3 is the login of the next user...
     USERLISTSTRING+="$UserName"
     USERLISTSTRING+=$'\n'
+    USERLISTSTRING+="$UserDisplayName"
+    USERLISTSTRING+=$'\n'
   done < <(echo "$USERLIST")
   #Append a default value with the display string to enter the username with -1. When selected, this tells the script to bring up a dialog to allow the user to enter a manual name
-  USERLISTSTRING+="Enter User Name..."
-  USERLISTSTRING+=$'\n'
   USERLISTSTRING+="-1"
+  USERLISTSTRING+=$'\n'
+  USERLISTSTRING+="Enter User Name..."
 
+  #Convert the data to an array
+  IFS=$'\n'
+  USERLISTSTRING=($USERLISTSTRING)
+  unset IFS
 
   #present the list of the users to the system
-  LOGINUSER=$(echo "$USERLISTSTRING" | $ZENITYCMD --title="Username" --height=450 --list --text "Select a user from the list to log into."  --column users --column usernames --hide-column 2 --print-column 2 --hide-header --separator="\n" 2>/dev/null )
+  LOGINUSER=$($KDIALOGCMD --title="Username" --menu "Select a user from the list to log into." -- "${USERLISTSTRING[@]}" 2>/dev/null )
   CancelOrOK=$?
   export LOGINUSER
-  #Abort if the user selected cancel on the zenity dialog
+  #Abort if the user selected cancel on the dialog
   if [[ $CancelOrOK != 0 ]]
   then
     exit 0
@@ -2358,9 +2345,9 @@
   #if there is no user selected or if the LOGINUSER is -1, as in the user opted to enter a username manually prompt for the username with a text dialog
   if [[ -z $LOGINUSER || $LOGINUSER == -1 ]]
   then
-    LOGINUSER=$($ZENITYCMD --title="Username" --entry --text="Enter the User Name:" 2>/dev/null)
+    LOGINUSER=$($KDIALOGCMD --title="Username" --inputbox "Enter the User Name:" 2>/dev/null)
     CancelOrOK=$?
-    #Abort if the user selected cancel on the zenity dialog
+    #Abort if the user selected cancel on the dialog
     if [[ $CancelOrOK != 0 ]]
     then
       exit 0
@@ -2391,9 +2378,9 @@
       passwordresult=0
     else
       #Prompt for the users password
-      USERPASSWORD=$( $ZENITYCMD --title="Password" --password --text "Type Password for $LOGINUSER" 2>/dev/null )
+      USERPASSWORD=$( $KDIALOGCMD --title="Password" --password "Type Password for $LOGINUSER" 2>/dev/null )
       CancelOrOK=$?
-      #Abort if the user selected cancel on the zenity dialog
+      #Abort if the user selected cancel on the dialog
       if [[ $CancelOrOK != 0 ]]
       then
         exit 0
@@ -2407,7 +2394,7 @@
     fi
     if [[ $passwordresult != 0 ]]
     then
-      $ZENITYCMD --warning --text="Invalid password for $LOGINUSER, or username invalid" 2>/dev/null
+      $KDIALOGCMD --title="Login Failed" --msgbox "Invalid password for $LOGINUSER, or username invalid" 2>/dev/null
       WriteToLog "Invalid password for $LOGINUSER, or username invalid"
     else
       export USERHOME=$(eval echo ~$LOGINUSER)
@@ -2430,13 +2417,13 @@
       StartUserSession "$SessionTTY" "$1" "$LOGINUSER" 0 "$WSESSIONDATA" &
     fi
   else
-    $ZENITYCMD --warning --text="Not Enough TTYs!" 2>/dev/null
+    $KDIALOGCMD --title="System Error" --error "Not Enough TTYs!" 2>/dev/null
     WriteToLog "Not Enough TTYs for user login!"
   fi
 }
 
 
-#This function is called when the loginmanager_control FIFO recives the command "Leave". It brings up a zenity dialog for handling power options for the computer.
+#This function is called when the loginmanager_control FIFO recives the command "Leave". It brings up a dialog for handling power options for the computer.
 #It takes 1 optional argument, the seat to display the dialogs on
 function LeavePrompt
 {
@@ -2449,7 +2436,7 @@
     SessionPromptThreshold=1
   else
     SetCommonConfig
-    ZENITYCMD=zenity
+    KDIALOGCMD=kdialog
     #Can't rely on SeatMultiSessionStatuses here, since this can be used outside of the service.
     SeatCanMultiSession=$(loginctl show-seat $XDG_SEAT -p CanMultiSession --value)
     SessionPromptThreshold=2
@@ -2488,7 +2475,7 @@
   unset RunningSessions
   unset RunningLoginManagerDisplaySessions
 
-  #The first line is sent to the zenity dialog is the actual command it sets the ACTION variable to be. the second line is what it appears as in the Zenity dialog for the user.
+  #The first line is sent to the dialog is the actual command it sets the ACTION variable to be. the second line is what it appears as in the dialog for the user.
   CanPowerOff=$(dbus-send --print-reply --system --dest=org.freedesktop.login1 /org/freedesktop/login1 org.freedesktop.login1.Manager.CanPowerOff)
   if [[ $CanPowerOff =~ " \"yes\"" ]]
   then
@@ -2544,9 +2531,15 @@
     ACTIONSTRING+=$'\n'
     ACTIONSTRING+="Hibernate Computer"
   fi
-  ACTION=$(echo "$ACTIONSTRING" | $ZENITYCMD --title="Leave..." --height=450 --list  --hide-header --text="What do you want to do?" --separator="\n" --column 'action' --column 'useraction' --hide-column 1 --print-column=1  2>/dev/null )
+
+  IFS=$'\n'
+  ACTIONSTRING=($ACTIONSTRING)
+  unset IFS
+
+  ACTION=$($KDIALOGCMD --title="Leave..." --menu "What do you want to do?" -- "${ACTIONSTRING[@]}" 2>/dev/null )
+
   CancelOrOK=$?
-  #Abort if the user selected cancel on the zenity dialog
+  #Abort if the user selected cancel on the dialog
   if [[ $CancelOrOK != 0 ]]
   then
     exit 0
@@ -2560,7 +2553,7 @@
   fi
 
   #Prompt the user if they are sure they want to execute the selected action
-  $ZENITYCMD --width=450 --title="Leave..." --question --text="Are you sure you want to ${ACTION}? $SESSIONINFO" 2>/dev/null
+  $KDIALOGCMD --title="Leave..." --yesno "Are you sure you want to ${ACTION}? $SESSIONINFO" 2>/dev/null
   CONFIRM=$?
   if [[ $CONFIRM != 0 ]]
   then
@@ -2609,7 +2602,7 @@
   #if the selected action failed, tell the user
   if [[ $RESULT != 0 ]]
   then
-    $ZENITYCMD --title="Failure" --warning --text="Failed to $ACTION" 2>/dev/null
+    $KDIALOGCMD --title="Failure" --error "Failed to $ACTION" 2>/dev/null
   fi
 }
 
@@ -2997,7 +2990,7 @@
   #Wait until the loginmanagerdisplay becomes availible
   DisplayServerWait "$1"
   ChooseSessionType "$1" "$AUTOLOGINUSER" 1
-  #Abort if the user selected cancel on the zenity dialog
+  #Abort if the user selected cancel on the dialog
   CancelOrOK=$?
   if [[ $CancelOrOK != 0 ]]
   then
@@ -3022,7 +3015,7 @@
     fi
     StartUserSession "$AutoSessionTTY" "$1" "$AUTOLOGINUSER" 1 "$WSESSIONDATA" &
   else
-    $ZENITYCMD --warning --text="Not Enough TTYs for autologin!" 2>/dev/null
+    $KDIALOGCMD --title="System Error" --error "Not Enough TTYs!" 2>/dev/null
     WriteToLog "Not Enough TTYs for autologin!"
   fi
 }
@@ -3056,19 +3049,24 @@
   SetFallbackEnvironmentVariables $1
   #wait for the loginmanagerdisplay
   DisplayServerWait "$1"
-  ActionMenu="Login\nLogin...\n"
+  ActionMenu="Login"$'\n'"Login..."$'\n'
   SeatCanMultiSession=$(loginctl show-seat "$1" -p CanMultiSession --value 2>/dev/null)
   if [[ $ONLY_CAN_MULTISESSION_SESSION_SWITCH == 0 || $SeatCanMultiSession == "yes" ]]
   then
-    ActionMenu+="Switch\nSwitch User...\n"
+    ActionMenu+="Switch"$'\n'"Switch User..."$'\n'
   fi
-  ActionMenu+="Leave\nShutdown...\nInfo\nHelp..."
+  ActionMenu+="Leave"$'\n'"Shutdown..."$'\n'"Info"$'\n'"Help..."
   if [[ $wlmdebuginsecure == 1 ]]
   then
-    ActionMenu+="\nDebug\nRoot Terminal..."
+    ActionMenu+=$'\n'"Debug"$'\n'"Root Terminal..."
   fi
-  USERACTION=$(echo -e "$ActionMenu"|$ZENITYCMD  --title="Login Manager" --height=300 --hide-column 1 --separator="\n" --list --column action --column display --hide-header --text "Select an Action"  2>> /var/log/waylandloginmanager/loginmanagerdisplays/loginmanagerdisplay_"$SeatFileName".log; exit ${PIPESTATUS[1]})
-ZENITYSTATUS=$?
+
+  IFS=$'\n'
+  ActionMenu=($ActionMenu)
+  unset IFS
+
+  USERACTION=$($KDIALOGCMD --title="Login Manager" --menu "Select an Action" -- "${ActionMenu[@]}"  2>> /var/log/waylandloginmanager/loginmanagerdisplays/loginmanagerdisplay_"$SeatFileName".log)
+  DIALOGSTATUS=$?
   #If the useraction exists (the user did not click cancel), then send the command to the loginmanager_control
   if [[ ! -z $USERACTION ]]
   then
@@ -3076,12 +3074,12 @@
   else
     $SLEEPCMD .1
   fi
-  #if zenity segfaults, then something is wrong. wait longer
-  if [[ $ZENITYSTATUS == 139 ]]
+  #if the dialog program segfaults, then something is wrong. wait longer
+  if [[ $DIALOGSTATUS == 139 ]]
   then
     $SLEEPCMD 5
   fi
-  exit $ZENITYSTATUS
+  exit $DIALOGSTATUS
 }
 
 #This part of the script is constantly running. It reads what the user send to loginmanager_control running as a service account, and filters out only valid commands to loginmanager_listener, and only sends a limited number of lines.
@@ -3415,9 +3413,9 @@
       fi
     done
 
-    if [[ $ZenityCommandFoundResult != 0 ]]
+    if [[ $KdialogCommandFoundResult != 0 ]]
     then
-      WriteToLog "zenity is not installed"
+      WriteToLog "kdialog is not installed"
       UIPromptAttemptsMaxedOut=1
     fi
 
