diff --git a/libweston/compositor.c b/libweston/compositor.c
index 2d864463..2dd07da9 100644
--- a/libweston/compositor.c
+++ b/libweston/compositor.c
@@ -7669,7 +7669,7 @@ weston_output_enable(struct weston_output *output)
 	}
 
 	weston_compositor_add_output(output->compositor, output);
-	weston_output_damage(output);
+	output->full_repaint_needed = true;
 
 	head_names = weston_output_create_heads_string(output);
 	weston_log("Output '%s' enabled with head(s) %s\n",
