https://gitlab.freedesktop.org/wayland/weston/-/merge_requests/16

diff --git a/ivi-shell/ivi-shell.c b/ivi-shell/ivi-shell.c
index 811a83ebc32d5e7e89342c101a234448bf533923..9df63c5f74380115fa3be62a2347c80d38dc4b45 100644
--- a/ivi-shell/ivi-shell.c
+++ b/ivi-shell/ivi-shell.c
@@ -648,6 +648,7 @@ wet_shell_init(struct weston_compositor *compositor,
 
 	ivi_layout_init_with_compositor(compositor);
 	shell_add_bindings(compositor, shell);
+	weston_compositor_damage_all(compositor);
 
 	return IVI_SUCCEEDED;
 
diff --git a/libweston/compositor.c b/libweston/compositor.c
index 6e9653721a868d0c98479df648192ff85738e8fd..34d8483c3a69ab368f3dec213e772f98ba6de9c4 100644
--- a/libweston/compositor.c
+++ b/libweston/compositor.c
@@ -6471,7 +6471,9 @@ weston_output_enable(struct weston_output *output)
 	weston_output_init_zoom(output);
 
 	weston_output_init_geometry(output, x, y);
-	weston_output_damage(output);
+	pixman_region32_union(&c->primary_plane.damage,
+			      &c->primary_plane.damage,
+			      &output->region);
 
 	wl_list_init(&output->animation_list);
 	wl_list_init(&output->feedback_list);

