diff --git a/libweston/compositor.c b/libweston/compositor.c
index 30913cbc..5f5d34a8 100644
--- a/libweston/compositor.c
+++ b/libweston/compositor.c
@@ -7325,6 +7325,7 @@ weston_output_placement_ok(struct weston_output *output)
 WL_EXPORT int
 weston_output_enable(struct weston_output *output)
 {
+	struct weston_compositor *c = output->compositor;
 	struct weston_head *head;
 	char *head_names;
 
@@ -7371,7 +7372,9 @@ weston_output_enable(struct weston_output *output)
 	if (!weston_output_placement_ok(output))
 		return -1;
 
-	weston_output_damage(output);
+	pixman_region32_union(&c->primary_plane.damage,
+			      &c->primary_plane.damage,
+			      &output->region);
 
 	wl_list_init(&output->animation_list);
 	wl_list_init(&output->feedback_list);
