diff --git a/ivi-shell/ivi-shell.c b/ivi-shell/ivi-shell.c
index eb6bbbe7..9f7d12b6 100644
--- a/ivi-shell/ivi-shell.c
+++ b/ivi-shell/ivi-shell.c
@@ -650,6 +650,7 @@ wet_shell_init(struct weston_compositor *compositor,
 
 	ivi_layout_init_with_compositor(compositor);
 	shell_add_bindings(compositor, shell);
+	weston_compositor_damage_all(compositor);
 
 	return IVI_SUCCEEDED;
 
diff --git a/libweston/compositor.c b/libweston/compositor.c
index 30913cbc..4dfb0f7d 100644
--- a/libweston/compositor.c
+++ b/libweston/compositor.c
@@ -7371,7 +7371,9 @@ weston_output_enable(struct weston_output *output)
 	if (!weston_output_placement_ok(output))
 		return -1;
 
-	weston_output_damage(output);
+	pixman_region32_union(&c->primary_plane.damage,
+			      &c->primary_plane.damage,
+			      &output->region);
 
 	wl_list_init(&output->animation_list);
 	wl_list_init(&output->feedback_list);
