diff --git a/src/compositor-drm.c b/src/compositor-drm.c
index 07b83a7..ab09c6c 100644
--- a/src/compositor-drm.c
+++ b/src/compositor-drm.c
@@ -1616,6 +1616,7 @@ drm_output_init_egl(struct drm_output *output, struct drm_compositor *ec)
 		weston_log("cursor buffers unavailable, using gl cursors\n");
 		ec->cursors_are_broken = 1;
 	}
+	ec->cursors_are_broken = 1;
 
 	return 0;
 }







diff --git a/src/compositor-fbdev.c b/src/compositor-fbdev.c
index 138aaab..7ebbe92 100644
--- a/src/compositor-fbdev.c
+++ b/src/compositor-fbdev.c
@@ -726,11 +726,14 @@ fbdev_output_reenable(struct fbdev_compositor *compositor,
 	weston_log("Re-enabling fbdev output.\n");
 
 	/* Create the frame buffer. */
-	fb_fd = fbdev_frame_buffer_open(output, output->device,
+	fb_fd = -1;
+	while (fb_fd < 0) {
+		usleep(10000);
+		fb_fd = fbdev_frame_buffer_open(output, output->device,
 	                                &new_screen_info);
-	if (fb_fd < 0) {
-		weston_log("Creating frame buffer failed.\n");
-		goto err;
+		if (fb_fd < 0) {
+		  weston_log("Creating frame buffer failed.\n");
+		}
 	}
 
 	/* Check whether the frame buffer details have changed since we were
