diff --git a/config.mk b/config.mk
index da93c6e..9ed1f77 100644
--- a/config.mk
+++ b/config.mk
@@ -1,5 +1,6 @@
 # wld: config.mk
 
+PREFIX              = /opt
 CC                  = gcc
 CFLAGS              = -pipe
 
@@ -9,7 +10,7 @@ ENABLE_PIXMAN       = 1
 ENABLE_DRM          = 1
 ENABLE_WAYLAND      = 1
 
-DRM_DRIVERS         = intel nouveau
+#DRM_DRIVERS         = intel nouveau
 WAYLAND_INTERFACES  = shm
 
 ifeq ($(ENABLE_DRM),1)
