diff --git a/CMakeLists.txt b/CMakeLists.txt
index bff3a4f..cfed0f8 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -56,7 +56,7 @@ if (NOT QT5_BUILD)
       ${CMAKE_SOURCE_DIR}/libinqt5/src/
       ${CMAKE_BINARY_DIR}/libinqt5/src/
   )
-  set(kwindowsystem_LIBRARY kwindowsystem)
+  
 endif()
 add_subdirectory( libqtmimetypes )
 add_subdirectory( libkdeqt5staging )
@@ -320,12 +320,11 @@ configure_file(kdemacros.h.cmake ${CMAKE_CURRENT_BINARY_DIR}/kdemacros.h )
 check_library_exists(nsl gethostbyname "" HAVE_NSL_LIBRARY)
 check_library_exists(socket connect "" HAVE_SOCKET_LIBRARY)
 
-# TODO: Port away from this:
-if (QT5_BUILD)
+
   add_definitions(-DKDE_NO_WINDOWSYSTEM)
   include_directories(kwindowsystem_dummy)
   add_definitions(-DKDE_NO_PHONON)
-endif()
+
 
 ################# list the subdirectories #################
 
