diff --git a/src/plugins/platforms/wayland/gl_integration/xcomposite_egl/qwaylandxcompositeeglcontext.h b/src/plugins/platforms/wayland/gl_integration/xcomposite_egl/qwaylandxcompositeeglcontext.h
index 3f6e1a1..a7d83d2 100644
--- a/src/plugins/platforms/wayland/gl_integration/xcomposite_egl/qwaylandxcompositeeglcontext.h
+++ b/src/plugins/platforms/wayland/gl_integration/xcomposite_egl/qwaylandxcompositeeglcontext.h
@@ -76,6 +76,7 @@ private:
     static const struct wl_callback_listener sync_callback_listener;
     static void sync_function(void *data, wl_callback *wl_callback, uint32_t time);
     bool mWaitingForSync;
+    bool mWaitingForSyncCallback;
 };
 
 #endif // QWAYLANDXCOMPOSITEEGLCONTEXT_H
diff --git a/src/plugins/platforms/wayland/wayland.pro b/src/plugins/platforms/wayland/wayland.pro
index 161886d..7713083 100644
--- a/src/plugins/platforms/wayland/wayland.pro
+++ b/src/plugins/platforms/wayland/wayland.pro
@@ -39,7 +39,7 @@ HEADERS =   qwaylandintegration.h \
             qwaylanddatasource.h \
             qwaylandshell.h \
             qwaylandshellsurface.h
-
+LIBS += -lX11
 INCLUDEPATH += $$QMAKE_INCDIR_WAYLAND
 LIBS += $$QMAKE_LIBS_WAYLAND
 QMAKE_CXXFLAGS += $$QMAKE_CFLAGS_WAYLAND
