diff --git a/src/weston-launch.c b/src/weston-launch.c
index 36f9c6b..7238128 100644
--- a/src/weston-launch.c
+++ b/src/weston-launch.c
@@ -109,7 +109,7 @@ weston_launch_allowed(struct weston_launch *wl)
        if (getuid() == 0)
                return 1;
 
-       gr = getgrnam("weston-launch");
+       gr = getgrnam("plugdev");
        if (gr) {
                groups = read_groups();
                if (groups) {
@@ -550,7 +550,7 @@ main(int argc, char *argv[])
 #else
                      " - enable systemd session support for weston-launch.\n"
 #endif
-                     " - or add yourself to the 'weston-launch' group.");
+                     " - or add yourself to the 'plugdev' group.");
 
        if (setup_tty(&wl, tty) < 0)
                return 1;