---
 src/radeon_kms.c |    8 ++++++--
 1 files changed, 6 insertions(+), 2 deletions(-)

diff --git a/src/radeon_kms.c b/src/radeon_kms.c
index 5e36d91..db3de74 100644
--- a/src/radeon_kms.c
+++ b/src/radeon_kms.c
@@ -797,13 +797,17 @@ Bool RADEONPreInit_KMS(ScrnInfoPtr pScrn, int flags)
 
 #ifdef XORG_WAYLAND
 	{
-		xf86LoadSubModule(pScrn, "xwayland");
-		info->xwl_screen = xwl_screen_pre_init(pScrn, 0, &xwl_driver);
+		info->xwl_screen = xwl_screen_create();
 		if (!info->xwl_screen) {
 			xf86DrvMsg(pScrn->scrnIndex, X_ERROR, "Failed to initialize xwayland.\n");
 			return FALSE;
 		}
 
+		if (!xwl_screen_pre_init(pScrn, info->xwl_screen, 0, &xwl_driver)) {
+			xf86DrvMsg(pScrn->scrnIndex, X_ERROR, "Failed to pre-init xwayland screen\n");
+			xwl_screen_destroy(info->xwl_screen);
+		}
+
 		info->dri2.drm_fd = xwl_screen_get_drm_fd(info->xwl_screen);
 		info->drmmode.fd = info->dri2.drm_fd;
 		info->dri->drmFD = info->dri2.drm_fd;
-- 
1.7.5.4