diff --git a/kf5/itemviews/CMakeLists.txt b/kf5/itemviews/CMakeLists.txt
index ae247a4..cbbd8c7 100644
--- a/kf5/itemviews/CMakeLists.txt
+++ b/kf5/itemviews/CMakeLists.txt
@@ -18,7 +18,7 @@ include(FeatureSummary)
 
 include(ECMSetupVersion)
 
-ecm_setup_version(5 0 0 VARIABLE_PREFIX ITEMVIEWS
+ecm_setup_version(5 VARIABLE_PREFIX ITEMVIEWS
                         VERSION_HEADER "${CMAKE_CURRENT_BINARY_DIR}/itemviews_version.h"
                         PACKAGE_VERSION_FILE "${CMAKE_CURRENT_BINARY_DIR}/ItemViewsConfigVersion.cmake")
 
diff --git a/kf5/kcodecs/CMakeLists.txt b/kf5/kcodecs/CMakeLists.txt
index f60431d..d674bdd 100644
--- a/kf5/kcodecs/CMakeLists.txt
+++ b/kf5/kcodecs/CMakeLists.txt
@@ -14,7 +14,7 @@ include(FeatureSummary)
 include(CMakePackageConfigHelpers)
 include(ECMSetupVersion)
 
-ecm_setup_version(5 0 0 VARIABLE_PREFIX KCODECS
+ecm_setup_version(5 VARIABLE_PREFIX KCODECS
                         VERSION_HEADER "${CMAKE_CURRENT_BINARY_DIR}/kcodecs_version.h"
                         PACKAGE_VERSION_FILE "${CMAKE_CURRENT_BINARY_DIR}/KCodecsConfigVersion.cmake")
 
diff --git a/kf5/kconfig/CMakeLists.txt b/kf5/kconfig/CMakeLists.txt
index 685c4d3..237f567 100644
--- a/kf5/kconfig/CMakeLists.txt
+++ b/kf5/kconfig/CMakeLists.txt
@@ -15,7 +15,7 @@ endif()
 include(FeatureSummary)
 
 include(ECMSetupVersion)
-ecm_setup_version(5 0 0 VARIABLE_PREFIX KCONFIG
+ecm_setup_version(5 VARIABLE_PREFIX KCONFIG
                         VERSION_HEADER "${CMAKE_CURRENT_BINARY_DIR}/kconfig_version.h"
                         PACKAGE_VERSION_FILE "${CMAKE_CURRENT_BINARY_DIR}/KConfigConfigVersion.cmake")
 
diff --git a/kf5/kconfigwidgets/CMakeLists.txt b/kf5/kconfigwidgets/CMakeLists.txt
index 5e563d6..3b72389 100644
--- a/kf5/kconfigwidgets/CMakeLists.txt
+++ b/kf5/kconfigwidgets/CMakeLists.txt
@@ -26,7 +26,7 @@ include(FeatureSummary)
 
 include(ECMSetupVersion)
 
-ecm_setup_version(5 0 0 VARIABLE_PREFIX KCONFIGWIDGETS
+ecm_setup_version(5 VARIABLE_PREFIX KCONFIGWIDGETS
                         VERSION_HEADER "${CMAKE_CURRENT_BINARY_DIR}/kconfigwidgets_version.h"
                         PACKAGE_VERSION_FILE "${CMAKE_CURRENT_BINARY_DIR}/KConfigWidgetsConfigVersion.cmake")
 
diff --git a/kf5/kcoreaddons/CMakeLists.txt b/kf5/kcoreaddons/CMakeLists.txt
index 4d4e8a0..ea40ef9 100644
--- a/kf5/kcoreaddons/CMakeLists.txt
+++ b/kf5/kcoreaddons/CMakeLists.txt
@@ -29,7 +29,7 @@ set(HAVE_SYS_INOTIFY_H ${SYS_INOTIFY_H_FOUND})
 # Generate io/config-kdirwatch.h
 configure_file(src/io/config-kdirwatch.h.cmake ${CMAKE_CURRENT_BINARY_DIR}/src/io/config-kdirwatch.h)
 
-ecm_setup_version(5 0 0 VARIABLE_PREFIX KCOREADDONS
+ecm_setup_version(5 VARIABLE_PREFIX KCOREADDONS
                         VERSION_HEADER "${CMAKE_CURRENT_BINARY_DIR}/kcoreaddons_version.h"
                         PACKAGE_VERSION_FILE "${CMAKE_CURRENT_BINARY_DIR}/KCoreAddonsConfigVersion.cmake")
 
diff --git a/kf5/ki18n/CMakeLists.txt b/kf5/ki18n/CMakeLists.txt
index c31839f..90d942e 100644
--- a/kf5/ki18n/CMakeLists.txt
+++ b/kf5/ki18n/CMakeLists.txt
@@ -14,7 +14,7 @@ find_package(KF5 5.0.0 REQUIRED MODULE COMPONENTS CMake Compiler InstallDirs)
 
 find_package(Qt5Test 5.1.0 REQUIRED NO_MODULE)
 
-ecm_setup_version(5 0 0 VARIABLE_PREFIX KI18N
+ecm_setup_version(5 VARIABLE_PREFIX KI18N
                         VERSION_HEADER "${CMAKE_CURRENT_BINARY_DIR}/ki18n_version.h"
                         PACKAGE_VERSION_FILE "${CMAKE_CURRENT_BINARY_DIR}/KI18nConfigVersion.cmake")
 
diff --git a/kf5/kservice/CMakeLists.txt b/kf5/kservice/CMakeLists.txt
index 29482d3..a2f04f2 100644
--- a/kf5/kservice/CMakeLists.txt
+++ b/kf5/kservice/CMakeLists.txt
@@ -2,7 +2,7 @@ project(KService)
 
 include(ECMSetupVersion)
 
-ecm_setup_version(5 0 0 VARIABLE_PREFIX KSERVICE
+ecm_setup_version(5 VARIABLE_PREFIX KSERVICE
                         VERSION_HEADER "${CMAKE_CURRENT_BINARY_DIR}/kservice_version.h"
                         PACKAGE_VERSION_FILE "${CMAKE_CURRENT_BINARY_DIR}/KServiceConfigVersion.cmake")
 
diff --git a/kf5/kwidgetsaddons/CMakeLists.txt b/kf5/kwidgetsaddons/CMakeLists.txt
index 7ec5776..c37ba6a 100644
--- a/kf5/kwidgetsaddons/CMakeLists.txt
+++ b/kf5/kwidgetsaddons/CMakeLists.txt
@@ -15,7 +15,7 @@ include(FeatureSummary)
 include(CMakePackageConfigHelpers)
 include(ECMSetupVersion)
 
-ecm_setup_version(5 0 0 VARIABLE_PREFIX KWIDGETSADDONS
+ecm_setup_version(5 VARIABLE_PREFIX KWIDGETSADDONS
                         VERSION_HEADER "${CMAKE_CURRENT_BINARY_DIR}/kwidgetsaddons_version.h"
                         PACKAGE_VERSION_FILE "${CMAKE_CURRENT_BINARY_DIR}/KWidgetsAddonsConfigVersion.cmake")
 
diff --git a/kf5/xmlgui/CMakeLists.txt b/kf5/xmlgui/CMakeLists.txt
index f24a08d..34a5f56 100644
--- a/kf5/xmlgui/CMakeLists.txt
+++ b/kf5/xmlgui/CMakeLists.txt
@@ -14,7 +14,7 @@ include(FeatureSummary)
 
 include(ECMSetupVersion)
 
-ecm_setup_version(5 0 0 VARIABLE_PREFIX XMLGUI
+ecm_setup_version(5 VARIABLE_PREFIX XMLGUI
                         VERSION_HEADER "${CMAKE_CURRENT_BINARY_DIR}/xmlgui_version.h"
                         PACKAGE_VERSION_FILE "${CMAKE_CURRENT_BINARY_DIR}/XmlGuiConfigVersion.cmake")
