diff --git a/kwin/killer/CMakeLists.txt b/kwin/killer/CMakeLists.txt
index db7855d..d6bb95f 100644
--- a/kwin/killer/CMakeLists.txt
+++ b/kwin/killer/CMakeLists.txt
@@ -4,7 +4,7 @@ set(kwin_killer_helper_SRCS killer.cpp )
 
 
 add_executable(kwin_killer_helper ${kwin_killer_helper_SRCS})
-set_target_properties(kwin_killer_helper PROPERTIES COMPILE_FLAGS ${Qt5Widgets_EXECUTABLE_COMPILE_FLAGS})
+#set_target_properties(kwin_killer_helper PROPERTIES COMPILE_FLAGS ${Qt5Widgets_EXECUTABLE_COMPILE_FLAGS})
 
 target_link_libraries(kwin_killer_helper
     Qt5::Widgets
