diff --git a/CMakeLists.txt b/CMakeLists.txt
index fb5a3d6..37dc091 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -137,7 +137,7 @@ include_directories (${CMAKE_SOURCE_DIR} ${CMAKE_CURRENT_BINARY_DIR} ${KDE4_INCL
 # Background processes
 # add_subdirectory(kpasswdserver)
 # add_subdirectory(kdontchangethehostname)
-add_subdirectory(kglobalaccel)
+# add_subdirectory(kglobalaccel)
 # add_subdirectory(knotify)
 add_subdirectory(ktimezoned)
 
diff --git a/kglobalaccel/CMakeLists.txt b/kglobalaccel/CMakeLists.txt
index c6dde6a..f817417 100644
--- a/kglobalaccel/CMakeLists.txt
+++ b/kglobalaccel/CMakeLists.txt
@@ -15,7 +15,7 @@ set(kglobalaccel_SRCS
 
 if (${XCB_XCB_FOUND} AND ${XCB_KEYSYMS_FOUND})
   include_directories(${XCB_INCLUDE_DIRS})
-  set( kglobalaccel_SRCS ${kglobalaccel_SRCS} kglobalaccel_x11.cpp )
+  #set( kglobalaccel_SRCS ${kglobalaccel_SRCS} kglobalaccel_x11.cpp )
 endif()
 if ( Q_WS_MAC )
   set( kglobalaccel_SRCS ${kglobalaccel_SRCS} kglobalaccel_mac.cpp )
