diff --git a/src/wlt_toolkit.c b/src/wlt_toolkit.c
index 18a0233..0b0278e 100644
--- a/src/wlt_toolkit.c
+++ b/src/wlt_toolkit.c
@@ -1295,6 +1295,8 @@ static void wlt_window_do_redraw(struct wlt_window *wnd,
 			y = 0;
 		wl_surface_attach(wnd->w_surface, wnd->w_buffer, x, y);
 		wnd->resize_edges = 0;
+	} else {
+		wl_surface_attach(wnd->w_surface, wnd->w_buffer, 0, 0);
 	}
 
 	wl_surface_damage(wnd->w_surface, 0, 0, wnd->buffer.width,
 
