=== modified file 'CMakeLists.txt'
--- CMakeLists.txt	2012-03-29 15:47:30 +0000
+++ CMakeLists.txt	2012-07-09 16:33:05 +0000
@@ -44,14 +44,25 @@
     )
 add_dependencies(distcheck dist)
 
+find_package(Qt5Widgets QUIET)
+if (Qt5Widgets_FOUND)
+   find_package(Qt5DBus REQUIRED)
+   include_directories(${Qt5Widgets_INCLUDE_DIRS} ${Qt5DBus_INCLUDE_DIRS})
+   find_package(ECM 0.0.4 REQUIRED)
+   set(CMAKE_MODULE_PATH ${CMAKE_MODULE_PATH} ${ECM_MODULE_PATH})
+   find_package(Qt5Transitional REQUIRED Core)
+   set(CMAKE_AUTOMOC ON)
+   set(CMAKE_AUTOMOC_RELAXED_MODE ON)
+else()
 find_package(Qt4 REQUIRED)
-
 include_directories(
     ${QT_INCLUDE_DIR}
     ${QT_QTCORE_INCLUDE_DIR}
     ${QT_QTDBUS_INCLUDE_DIR}
     ${QT_QTGUI_INCLUDE_DIR}
     )
+endif()
+
 
 configure_file(dbusmenu-qt.pc.in ${CMAKE_BINARY_DIR}/dbusmenu-qt.pc @ONLY)
 
@@ -60,7 +71,9 @@
     )
 
 add_subdirectory(src)
+if (NOT Qt5Widgets_FOUND) # TODO port tests to Qt5
 add_subdirectory(tests)
+endif()
 add_subdirectory(tools)
 
 if(WITH_DOC)

=== modified file 'src/CMakeLists.txt'
--- src/CMakeLists.txt	2012-04-23 14:48:31 +0000
+++ src/CMakeLists.txt	2012-05-11 10:53:49 +0000
@@ -49,7 +49,9 @@
     ${CMAKE_BINARY_DIR}/src
     )
 
+if (NOT Qt5Widgets_FOUND)
 qt4_automoc(${dbusmenu_qt_SRCS})
+endif()
 
 qt4_add_dbus_adaptor(dbusmenu_qt_SRCS
     ${CMAKE_CURRENT_SOURCE_DIR}/com.canonical.dbusmenu.xml

