diff --git a/configure.ac b/configure.ac
index 62d06e0..bc0fb46 100644
--- a/configure.ac
+++ b/configure.ac
@@ -748,6 +748,20 @@ fi
 
 AM_CONDITIONAL(HAVE_SHARED_GLAPI, test "x$enable_shared_glapi" = xyes)
 
+AC_ARG_ENABLE([shared-gallium],
+    [AS_HELP_STRING([--enable-shared-gallium],
+        [Enable shared gallium core @<:@default=yes@:>@])],
+    [enable_shared_gallium="$enableval"],
+    [enable_shared_gallium=yes])
+
+SHARED_GALLIUM="0"
+if test "x$enable_shared_gallium" = xyes; then
+    SHARED_GALLIUM="1"
+fi
+AC_SUBST([SHARED_GALLIUM])
+AM_CONDITIONAL(HAVE_SHARED_GALLIUM, test $SHARED_GALLIUM = 1)
+
+
 dnl
 dnl Driver specific build directories
 dnl
@@ -879,6 +893,8 @@ AC_SUBST([GLESv2_LIB_DEPS])
 AC_SUBST([GLESv2_PC_LIB_PRIV])
 
 DRI_LIB_DEPS="\$(top_builddir)/src/mesa/libdricore/libdricore${VERSION}.la"
+MESAGALLIUM_LIBS="${DRI_LIB_DEPS}"
+AC_SUBST([MESAGALLIUM_LIBS], ${MESAGALLIUM_LIBS})
 
 AC_SUBST([HAVE_XF86VIDMODE])
 
diff --git a/src/gallium/auxiliary/Makefile.am b/src/gallium/auxiliary/Makefile.am
index f14279b..bfda3f8 100644
--- a/src/gallium/auxiliary/Makefile.am
+++ b/src/gallium/auxiliary/Makefile.am
@@ -3,14 +3,20 @@ AUTOMAKE_OPTIONS = subdir-objects
 include Makefile.sources
 include $(top_srcdir)/src/gallium/Automake.inc
 
-noinst_LTLIBRARIES = libgallium.la
 
 AM_CFLAGS = \
 	-I$(top_srcdir)/src/gallium/auxiliary/util \
-	$(GALLIUM_CFLAGS) \
-	$(VISIBILITY_CFLAGS)
+	$(GALLIUM_CFLAGS)
+
+AM_CXXFLAGS =
 
-AM_CXXFLAGS = $(VISIBILITY_CXXFLAGS)
+if HAVE_SHARED_GALLIUM
+lib_LTLIBRARIES = libgallium.la
+else
+noinst_LTLIBRARIES = libgallium.la
+AM_CFLAGS += $(VISIBILITY_CFLAGS)
+AM_CXXFLAGS += $(VISIBILITY_CXXFLAGS)
+endif
 
 libgallium_la_SOURCES = \
 	$(C_SOURCES) \
@@ -35,6 +41,8 @@ libgallium_la_SOURCES += \
 	$(GALLIVM_SOURCES) \
 	$(GALLIVM_CPP_SOURCES)
 
+libgallium_la_LIBADD = ../../mesa/libdricore/libmesagallium.la $(LLVM_LIBS) $(GALLIUM_DRI_LIB_DEPS)
+
 endif
 
 indices/u_indices_gen.c: $(srcdir)/indices/u_indices_gen.py
diff --git a/src/gallium/targets/dri-freedreno/Makefile.am b/src/gallium/targets/dri-freedreno/Makefile.am
index cfa1f37..cc93f1d 100644
--- a/src/gallium/targets/dri-freedreno/Makefile.am
+++ b/src/gallium/targets/dri-freedreno/Makefile.am
@@ -49,7 +49,7 @@ kgsl_dri_la_SOURCES = \
 kgsl_dri_la_LDFLAGS = -module -avoid-version -shared -no-undefined
 
 kgsl_dri_la_LIBADD = \
-	$(top_builddir)/src/mesa/libmesagallium.la \
+	@MESAGALLIUM_LIBS@ \
 	$(top_builddir)/src/gallium/auxiliary/libgallium.la \
 	$(top_builddir)/src/gallium/state_trackers/dri/drm/libdridrm.la \
 	$(top_builddir)/src/gallium/winsys/freedreno/drm/libfreedrenodrm.la \
diff --git a/src/gallium/targets/dri-i915/Makefile.am b/src/gallium/targets/dri-i915/Makefile.am
index ce6be78..ee700dc 100644
--- a/src/gallium/targets/dri-i915/Makefile.am
+++ b/src/gallium/targets/dri-i915/Makefile.am
@@ -49,7 +49,7 @@ i915_dri_la_SOURCES = \
 i915_dri_la_LDFLAGS = -module -avoid-version -shared -no-undefined
 
 i915_dri_la_LIBADD = \
-	$(top_builddir)/src/mesa/libmesagallium.la \
+	@MESAGALLIUM_LIBS@ \
 	$(top_builddir)/src/gallium/auxiliary/libgallium.la \
 	$(top_builddir)/src/gallium/state_trackers/dri/drm/libdridrm.la \
 	$(top_builddir)/src/gallium/winsys/i915/drm/libi915drm.la \
diff --git a/src/gallium/targets/dri-ilo/Makefile.am b/src/gallium/targets/dri-ilo/Makefile.am
index 7761f33..90d971a 100644
--- a/src/gallium/targets/dri-ilo/Makefile.am
+++ b/src/gallium/targets/dri-ilo/Makefile.am
@@ -50,7 +50,7 @@ ilo_dri_la_LDFLAGS = -module -avoid-version -shared -no-undefined \
 		     -rpath $(abs_builddir)
 
 ilo_dri_la_LIBADD = \
-	$(top_builddir)/src/mesa/libmesagallium.la \
+	@MESAGALLIUM_LIBS@ \
 	$(top_builddir)/src/gallium/auxiliary/libgallium.la \
 	$(top_builddir)/src/gallium/state_trackers/dri/drm/libdridrm.la \
 	$(top_builddir)/src/gallium/winsys/intel/drm/libintelwinsys.la \
diff --git a/src/gallium/targets/dri-nouveau/Makefile.am b/src/gallium/targets/dri-nouveau/Makefile.am
index 69ccf32..fef4c63 100644
--- a/src/gallium/targets/dri-nouveau/Makefile.am
+++ b/src/gallium/targets/dri-nouveau/Makefile.am
@@ -48,7 +48,7 @@ nouveau_dri_la_SOURCES = \
 nouveau_dri_la_LDFLAGS = -module -avoid-version -shared -no-undefined
 
 nouveau_dri_la_LIBADD = \
-	$(top_builddir)/src/mesa/libmesagallium.la \
+	@MESAGALLIUM_LIBS@ \
 	$(top_builddir)/src/gallium/auxiliary/libgallium.la \
 	$(top_builddir)/src/gallium/state_trackers/dri/drm/libdridrm.la \
 	$(top_builddir)/src/gallium/winsys/nouveau/drm/libnouveaudrm.la \
diff --git a/src/gallium/targets/dri-r300/Makefile.am b/src/gallium/targets/dri-r300/Makefile.am
index 8c0215d..6beb536 100644
--- a/src/gallium/targets/dri-r300/Makefile.am
+++ b/src/gallium/targets/dri-r300/Makefile.am
@@ -49,7 +49,7 @@ r300_dri_la_SOURCES = \
 r300_dri_la_LDFLAGS = -module -avoid-version -shared -no-undefined
 
 r300_dri_la_LIBADD = \
-	$(top_builddir)/src/mesa/libmesagallium.la \
+	@MESAGALLIUM_LIBS@ \
 	$(top_builddir)/src/gallium/auxiliary/libgallium.la \
 	$(top_builddir)/src/gallium/state_trackers/dri/drm/libdridrm.la \
 	$(top_builddir)/src/gallium/winsys/radeon/drm/libradeonwinsys.la \
diff --git a/src/gallium/targets/dri-r600/Makefile.am b/src/gallium/targets/dri-r600/Makefile.am
index 2b3524b..d40fb89 100644
--- a/src/gallium/targets/dri-r600/Makefile.am
+++ b/src/gallium/targets/dri-r600/Makefile.am
@@ -48,7 +48,7 @@ r600_dri_la_SOURCES = \
 r600_dri_la_LDFLAGS = -module -avoid-version -shared -no-undefined
 
 r600_dri_la_LIBADD = \
-	$(top_builddir)/src/mesa/libmesagallium.la \
+	@MESAGALLIUM_LIBS@ \
 	$(top_builddir)/src/gallium/auxiliary/libgallium.la \
 	$(top_builddir)/src/gallium/drivers/r600/libr600.la \
 	$(top_builddir)/src/gallium/state_trackers/dri/drm/libdridrm.la \
diff --git a/src/gallium/targets/dri-radeonsi/Makefile.am b/src/gallium/targets/dri-radeonsi/Makefile.am
index f7d87a6..87ab2aa 100644
--- a/src/gallium/targets/dri-radeonsi/Makefile.am
+++ b/src/gallium/targets/dri-radeonsi/Makefile.am
@@ -49,7 +49,7 @@ radeonsi_dri_la_SOURCES = \
 radeonsi_dri_la_LDFLAGS = -module -avoid-version -shared -no-undefined
 
 radeonsi_dri_la_LIBADD = \
-	$(top_builddir)/src/mesa/libmesagallium.la \
+	@MESAGALLIUM_LIBS@ \
 	$(top_builddir)/src/gallium/auxiliary/libgallium.la \
 	$(top_builddir)/src/gallium/drivers/radeonsi/libradeonsi.la \
 	$(top_builddir)/src/gallium/state_trackers/dri/drm/libdridrm.la \
diff --git a/src/gallium/targets/dri-swrast/Makefile.am b/src/gallium/targets/dri-swrast/Makefile.am
index 1104379..72d2401 100644
--- a/src/gallium/targets/dri-swrast/Makefile.am
+++ b/src/gallium/targets/dri-swrast/Makefile.am
@@ -49,7 +49,7 @@ swrast_dri_la_SOURCES = \
 swrast_dri_la_LDFLAGS = -module -avoid-version -shared -no-undefined
 
 swrast_dri_la_LIBADD = \
-	$(top_builddir)/src/mesa/libmesagallium.la \
+	@MESAGALLIUM_LIBS@ \
 	$(top_builddir)/src/gallium/auxiliary/libgallium.la \
 	$(top_builddir)/src/gallium/state_trackers/dri/sw/libdrisw.la \
 	$(top_builddir)/src/gallium/winsys/sw/dri/libswdri.la \
diff --git a/src/gallium/targets/dri-vmwgfx/Makefile.am b/src/gallium/targets/dri-vmwgfx/Makefile.am
index ca7df65..4384976 100644
--- a/src/gallium/targets/dri-vmwgfx/Makefile.am
+++ b/src/gallium/targets/dri-vmwgfx/Makefile.am
@@ -48,7 +48,7 @@ vmwgfx_dri_la_SOURCES = \
 vmwgfx_dri_la_LDFLAGS = -module -avoid-version -shared -no-undefined
 
 vmwgfx_dri_la_LIBADD = \
-	$(top_builddir)/src/mesa/libmesagallium.la \
+	@MESAGALLIUM_LIBS@ \
 	$(top_builddir)/src/gallium/auxiliary/libgallium.la \
 	$(top_builddir)/src/gallium/state_trackers/dri/drm/libdridrm.la \
 	$(top_builddir)/src/gallium/winsys/svga/drm/libsvgadrm.la \
diff --git a/src/gallium/targets/egl-static/Makefile.am b/src/gallium/targets/egl-static/Makefile.am
index 90178f6..c90d87d 100644
--- a/src/gallium/targets/egl-static/Makefile.am
+++ b/src/gallium/targets/egl-static/Makefile.am
@@ -103,7 +103,7 @@ AM_CPPFLAGS += \
 	-DFEATURE_GL=1
 
 egl_gallium_la_LIBADD += \
-	$(top_builddir)/src/mesa/libmesagallium.la
+	@MESAGALLIUM_LIBS@ \
 # make st/mesa built-in when there is a single glapi provider
 if HAVE_SHARED_GLAPI
 egl_gallium_la_LIBADD += \
diff --git a/src/mesa/libdricore/Makefile.am b/src/mesa/libdricore/Makefile.am
index 56ceeb7..ca2e728 100644
--- a/src/mesa/libdricore/Makefile.am
+++ b/src/mesa/libdricore/Makefile.am
@@ -41,8 +41,10 @@ libdricore@VERSION@_la_SOURCES = \
 libdricore@VERSION@_la_LDFLAGS = -version-number 1:0
 libdricore@VERSION@_la_LIBADD = \
         ../program/libdricore_program.la \
+	$(top_builddir)/src/mapi/shared-glapi/libglapi.la
         $()
 
+lib_LTLIBRARIES =
 if HAVE_X86_ASM
 libdricore@VERSION@_la_SOURCES += $(X86_FILES)
 AM_CPPFLAGS += \
@@ -75,6 +77,16 @@ all-local: libdricore@VERSION@.la
 	ln -sf libdricore@VERSION@.so.1 $(top_builddir)/$(LIB_DIR)/libdricore@VERSION@.so
 endif
 
+libmesagallium_la_SOURCES = \
+	$(STATETRACKER_FILES)
+libmesagallium_la_CFLAGS = @LLVM_CFLAGS@
+libmesagallium_la_CXXFLAGS = @LLVM_CXXFLAGS@
+libmesagallium_la_LIBADD = libdricore@VERSION@.la $(LLVM_LIBS)
+
+if HAVE_GALLIUM
+noinst_LTLIBRARIES = libmesagallium.la
+endif
+
 CLEANFILES = \
 	$(top_builddir)/$(LIB_DIR)/libdricore@VERSION@.so.1 \
 	$(top_builddir)/$(LIB_DIR)/libdricore@VERSION@.so
