diff --git a/src/gallium/state_trackers/egl/wayland/native_shm.c b/src/gallium/state_trackers/egl/wayland/native_shm.c
index a959237..2499677 100644
--- a/src/gallium/state_trackers/egl/wayland/native_shm.c
+++ b/src/gallium/state_trackers/egl/wayland/native_shm.c
@@ -163,7 +163,7 @@ wayland_shm_display_init_screen(struct native_display *ndpy)
       return FALSE;
 
    if (shmdpy->base.formats == 0)
-      wayland_roundtrip(shmdpy->base.dpy);
+      wayland_roundtrip(&shmdpy->base);
    if (shmdpy->base.formats == 0)
       return FALSE;
  
