diff --git a/src/shell/shell.cpp b/src/shell/shell.cpp
index 9766f6b..7b56ca7 100644
--- a/src/shell/shell.cpp
+++ b/src/shell/shell.cpp
@@ -156,7 +156,7 @@ void Shell::init()
         m_backgroundLayer.addSurface(blackSurface);
         m_blackSurfaces.push_back(blackSurface);
 
-        weston_surface *splashSurface = createBlackSurface(x, y, w, h);
+        weston_surface *splashSurface = createBlackSurface(x, y, 0, 0);
         m_splashLayer.addSurface(splashSurface);
         m_splash->addOutput(splashSurface);
     }
