diff --git a/mkspecs/features/cmake_functions.prf b/mkspecs/features/cmake_functions.prf
index f48ef43..01c3f7a 100644
--- a/mkspecs/features/cmake_functions.prf
+++ b/mkspecs/features/cmake_functions.prf
@@ -54,7 +54,7 @@ defineReplace(cmakeProcessLibs) {
     variable = $$1
     out =
     for(v, variable) {
-        if(!equals(v, -framework):!equals(v, -L.*)) {
+        if(!equals(v, -framework):!contains(v, -L.*)) {
             v ~= s,^-l,,
             v ~= s,^-lib,,
             v ~= s,.lib$,,