diff --git a/src/compositor/wayland_wrapper/wloutput.cpp b/src/compositor/wayland_wrapper/wloutput.cpp
index e00d8b8..9ea1c73 100644
--- a/src/compositor/wayland_wrapper/wloutput.cpp
+++ b/src/compositor/wayland_wrapper/wloutput.cpp
@@ -95,7 +95,7 @@ Output::Output(OutputGlobal *outputGlobal, wl_client *client, uint32_t version,
     Q_UNUSED(version);
     m_output_resource = wl_client_add_object(client,&wl_output_interface,0,id,this);
     wl_output_send_geometry(m_output_resource, 0, 0,
-                         m_output_global->size().width(), m_output_global->size().height(),0,"","");
+                         m_output_global->size().width(), m_output_global->size().height(),0,"","",0);
 
     wl_output_send_mode(m_output_resource, WL_OUTPUT_MODE_CURRENT|WL_OUTPUT_MODE_PREFERRED,
                         m_output_global->size().width(), m_output_global->size().height(), m_output_global->refreshRate());
diff --git a/src/plugins/platforms/wayland/qwaylanddisplay.cpp b/src/plugins/platforms/wayland/qwaylanddisplay.cpp
index 4d0d680..33341e5 100644
--- a/src/plugins/platforms/wayland/qwaylanddisplay.cpp
+++ b/src/plugins/platforms/wayland/qwaylanddisplay.cpp
@@ -223,7 +223,8 @@ void QWaylandDisplay::outputHandleGeometry(void *data,
                                            int32_t physicalWidth,
                                            int32_t physicalHeight,
                                            int subpixel,
-                                           const char *make, const char *model)
+                                           const char *make, const char *model,
+					   int transform)
 {
     Q_UNUSED(subpixel);
     Q_UNUSED(make);
diff --git a/src/plugins/platforms/wayland/qwaylanddisplay.h b/src/plugins/platforms/wayland/qwaylanddisplay.h
index a24b907..5126cfc 100644
--- a/src/plugins/platforms/wayland/qwaylanddisplay.h
+++ b/src/plugins/platforms/wayland/qwaylanddisplay.h
@@ -152,7 +152,8 @@ private:
                                      int32_t width, int32_t height,
                                      int subpixel,
                                      const char *make,
-                                     const char *model);
+                                     const char *model,
+				     int transform);
     static void mode(void *data,
                      struct wl_output *wl_output,
                      uint32_t flags,
