diff --git a/src/webtab/webpage.cpp b/src/webtab/webpage.cpp
index ca9affd..029d4cf 100644
--- a/src/webtab/webpage.cpp
+++ b/src/webtab/webpage.cpp
@@ -717,7 +717,7 @@ QString WebPage::errorPage(QNetworkReply *reply)
     QString title = i18n("There was a problem while loading the page");
 
     QString msg;
-
+/*
     // test to see if networking is enabled on the system
     if (Solid::Networking::status() != Solid::Networking::Connected)
     {
@@ -747,7 +747,7 @@ QString WebPage::errorPage(QNetworkReply *reply)
         
         return html;
     }
-
+*/
     QString errString = reply->errorString().toLower();
     if (errString.contains(QL1S("proxy")))
     {
