diff --git a/src/weston-launch.c b/src/weston-launch.c
index 5b03094..d64af11 100644
--- a/src/weston-launch.c
+++ b/src/weston-launch.c
@@ -116,7 +116,7 @@ weston_launch_allowed(struct weston_launch *wl)
 	if (getuid() == 0)
 		return 1;
 
-	gr = getgrnam("weston-launch");
+	gr = getgrnam("plugdev");
 	if (gr) {
 		groups = read_groups();
 		if (groups) {
@@ -514,7 +514,6 @@ setup_session(struct weston_launch *wl)
 	}
 
 	term = getenv("TERM");
-	clearenv();
 	setenv("TERM", term, 1);
 	setenv("USER", wl->pw->pw_name, 1);
 	setenv("LOGNAME", wl->pw->pw_name, 1);
@@ -645,7 +644,7 @@ main(int argc, char *argv[])
 #else
 		      " - enable systemd session support for weston-launch.\n"
 #endif
-		      " - or add yourself to the 'weston-launch' group.");
+		      " - or add yourself to the 'plugdev' group.");
 
 	if (setup_tty(&wl, tty) < 0)
 		exit(EXIT_FAILURE);
