diff --git a/src/weston-launch.c b/src/weston-launch.c
index 36f9c6b..7238128 100644
--- a/src/weston-launch.c
+++ b/src/weston-launch.c
@@ -109,7 +109,7 @@ weston_launch_allowed(struct weston_launch *wl)
 	if (getuid() == 0)
 		return 1;
 
-	gr = getgrnam("weston-launch");
+	gr = getgrnam("plugdev");
 	if (gr) {
 		groups = read_groups();
 		if (groups) {
@@ -550,7 +550,7 @@ main(int argc, char *argv[])
 #else
 		      " - enable systemd session support for weston-launch.\n"
 #endif
-		      " - or add yourself to the 'weston-launch' group.");
+		      " - or add yourself to the 'plugdev' group.");
 
 	if (setup_tty(&wl, tty) < 0)
 		return 1;






The --fullscreen option effectively does nothing if --width and --height aren't
manually set to the size of the screen. This handles the size setting and limits
the output count to one, for fullscreen.
---
 src/compositor-x11.c | 6 +++++-
 1 file changed, 5 insertions(+), 1 deletion(-)

diff --git a/src/compositor-x11.c b/src/compositor-x11.c
index c575f25..75de857 100644
--- a/src/compositor-x11.c
+++ b/src/compositor-x11.c
@@ -1206,6 +1206,10 @@ x11_compositor_create(struct wl_display *display,
 	count = option_count ? option_count : 1;
 
 	wl_list_for_each(o, &configured_output_list, link) {
+		if (fullscreen) {
+			o->width = c->screen->width_in_pixels;
+			o->height = c->screen->height_in_pixels;
+		}
 		output = x11_compositor_create_output(c, x, 0,
 						      option_width ? width :
 						      o->width,
@@ -1219,7 +1223,7 @@ x11_compositor_create(struct wl_display *display,
 		x = pixman_region32_extents(&output->base.region)->x2;
 
 		output_count++;
-		if (option_count && output_count >= option_count)
+		if (fullscreen || (option_count && output_count >= option_count))
 			break;
 	}
 
-- 
1.7.11.7

