diff --git a/src/weston-launch.c b/src/weston-launch.c
index 7264f7e..f689131 100644
--- a/src/weston-launch.c
+++ b/src/weston-launch.c
@@ -116,7 +116,7 @@ weston_launch_allowed(struct weston_launch *wl)
 	if (getuid() == 0)
 		return 1;
 
-	gr = getgrnam("weston-launch");
+	gr = getgrnam("plugdev");
 	if (gr) {
 		groups = read_groups();
 		if (groups) {
@@ -514,7 +514,7 @@ setup_session(struct weston_launch *wl)
 	}
 
 	term = getenv("TERM");
-	clearenv();
+	//clearenv();
 	if (term)
 		setenv("TERM", term, 1);
 	setenv("USER", wl->pw->pw_name, 1);
@@ -646,7 +646,7 @@ main(int argc, char *argv[])
 #else
 		      " - enable systemd session support for weston-launch.\n"
 #endif
-		      " - or add yourself to the 'weston-launch' group.");
+		      " - or add yourself to the 'plugdev' group.");
 
 	if (setup_tty(&wl, tty) < 0)
 		exit(EXIT_FAILURE);




diff --git a/src/compositor.c b/src/compositor.c
index 4540911..b249459 100644
--- a/src/compositor.c
+++ b/src/compositor.c
@@ -4711,13 +4711,6 @@ int main(int argc, char *argv[])
 	ec->idle_time = idle_time;
 	ec->default_pointer_grab = NULL;
 
-	for (i = 1; i < argc; i++)
-		weston_log("fatal: unhandled option: %s\n", argv[i]);
-	if (argc > 1) {
-		ret = EXIT_FAILURE;
-		goto out;
-	}
-
 	weston_compositor_log_capabilities(ec);
 
 	server_socket = getenv("WAYLAND_SERVER_SOCKET");
